// vscode/service-override/keybindings.js
import { __decorate, __param } from '../external/tslib/tslib.es6.js';
import { StandaloneServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices.js';
import { SyncDescriptor } from 'monaco-editor/esm/vs/platform/instantiation/common/descriptors.js';
import { WorkbenchKeybindingService } from '../vscode/src/vs/workbench/services/keybinding/browser/keybindingService.js';
import { IKeybindingService } from 'monaco-editor/esm/vs/platform/keybinding/common/keybinding.js';
import { VSBuffer } from 'monaco-editor/esm/vs/base/common/buffer.js';
import { IUserDataProfilesService } from 'monaco-editor/esm/vs/platform/userDataProfile/common/userDataProfile.js';
import { IKeyboardLayoutService } from '../vscode/src/vs/platform/keyboardLayout/common/keyboardLayout.js';
import { BrowserKeyboardLayoutService } from '../vscode/src/vs/workbench/services/keybinding/browser/keyboardLayoutService.js';
import { IFileService } from 'monaco-editor/esm/vs/platform/files/common/files.js';
import { ICommandService } from 'monaco-editor/esm/vs/platform/commands/common/commands.js';
import { CommandService } from '../vscode/src/vs/workbench/services/commands/common/commandService.js';
import { DisposableStore, toDisposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { IContextKeyService } from 'monaco-editor/esm/vs/platform/contextkey/common/contextkey.js';
import { IUriIdentityService } from 'monaco-editor/esm/vs/platform/uriIdentity/common/uriIdentity.js';
import { ITelemetryService } from 'monaco-editor/esm/vs/platform/telemetry/common/telemetry.js';
import { INotificationService } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
import { IUserDataProfileService } from '../vscode/src/vs/workbench/services/userDataProfile/common/userDataProfile.js';
import { IHostService } from '../vscode/src/vs/workbench/services/host/browser/host.js';
import { IExtensionService } from '../vscode/src/vs/workbench/services/extensions/common/extensions.js';
import { ILogService } from 'monaco-editor/esm/vs/platform/log/common/log.js';
import { WorkbenchContextKeysHandler } from '../vscode/src/vs/workbench/browser/contextkeys.js';
import { Schemas } from 'monaco-editor/esm/vs/base/common/network.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import getServiceOverride$1, { initFile } from './files.js';
import '../monaco.js';
import { onRenderWorkbench } from '../lifecycle.js';
import '../services.js';
import '../vscode/src/vs/workbench/browser/workbench.contribution.js';
import '../vscode/src/vs/workbench/contrib/keybindings/browser/keybindings.contribution.js';
import { IInstantiationService } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
const defaultUserKeybindindsFile = ( URI.from({ scheme: Schemas.vscodeUserData, path: '/User/keybindings.json' }));
async function initUserKeybindings(configurationJson, options, file = defaultUserKeybindindsFile) {
    await initFile(defaultUserKeybindindsFile.scheme, file, configurationJson, options);
}
async function updateUserKeybindings(keybindingsJson) {
    const userDataProfilesService = StandaloneServices.get(IUserDataProfilesService);
    await StandaloneServices.get(IFileService).writeFile(userDataProfilesService.defaultProfile.keybindingsResource, VSBuffer.fromString(keybindingsJson));
}
let DynamicWorkbenchKeybindingService = class DynamicWorkbenchKeybindingService extends WorkbenchKeybindingService {
    constructor(shouldUseGlobalKeybindings, contextKeyService, commandService, telemetryService, notificationService, userDataProfileService, hostService, extensionService, fileService, uriIdentityService, logService, keyboardLayoutService) {
        super(contextKeyService, commandService, telemetryService, notificationService, userDataProfileService, hostService, extensionService, fileService, uriIdentityService, logService, keyboardLayoutService);
        this.shouldUseGlobalKeybindings = shouldUseGlobalKeybindings;
        this.keybindingProviders = [];
    }
    registerKeybindingProvider(provider) {
        this.keybindingProviders.push(provider);
        this.updateResolver();
        const store = new DisposableStore();
        store.add(provider.onDidChangeKeybindings(() => {
            this.updateResolver();
        }));
        store.add(toDisposable(() => {
            const idx = this.keybindingProviders.indexOf(provider);
            if (idx >= 0) {
                this.keybindingProviders.splice(idx, 1);
                this.updateResolver();
            }
        }));
        return store;
    }
    _getResolver() {
        return super._getResolver();
    }
    _dispatch(e, target) {
        if (!this.shouldUseGlobalKeybindings()) {
            return false;
        }
        return super._dispatch(e, target);
    }
    getUserKeybindingItems() {
        return [...super.getUserKeybindingItems(), ...this.keybindingProviders.flatMap(provider => provider.provideKeybindings())];
    }
};
DynamicWorkbenchKeybindingService = __decorate([
    ( __param(1, IContextKeyService)),
    ( __param(2, ICommandService)),
    ( __param(3, ITelemetryService)),
    ( __param(4, INotificationService)),
    ( __param(5, IUserDataProfileService)),
    ( __param(6, IHostService)),
    ( __param(7, IExtensionService)),
    ( __param(8, IFileService)),
    ( __param(9, IUriIdentityService)),
    ( __param(10, ILogService)),
    ( __param(11, IKeyboardLayoutService))
], DynamicWorkbenchKeybindingService);
onRenderWorkbench((accessor) => {
    accessor.get(IInstantiationService).createInstance(WorkbenchContextKeysHandler);
});
function getServiceOverride({ shouldUseGlobalKeybindings = () => false } = {}) {
    return {
        ...getServiceOverride$1(),
        [( IKeybindingService.toString())]: new SyncDescriptor(DynamicWorkbenchKeybindingService, [shouldUseGlobalKeybindings], false),
        [( IKeyboardLayoutService.toString())]: new SyncDescriptor(BrowserKeyboardLayoutService, undefined, true),
        [( ICommandService.toString())]: new SyncDescriptor(CommandService, [], true)
    };
}
export { getServiceOverride as default, defaultUserKeybindindsFile, initUserKeybindings, updateUserKeybindings };
