// vscode/vscode/src/vs/workbench/services/remote/common/remoteExplorerService.js
import { createDecorator } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
const IRemoteExplorerService = ( createDecorator('remoteExplorerService'));
const TUNNEL_VIEW_ID = '~remote.forwardedPorts';
const PORT_AUTO_FORWARD_SETTING = 'remote.autoForwardPorts';
const PORT_AUTO_SOURCE_SETTING = 'remote.autoForwardPortsSource';
const PORT_AUTO_SOURCE_SETTING_OUTPUT = 'output';
var TunnelType;
( (function(TunnelType) {
    TunnelType["Candidate"] = "Candidate";
    TunnelType["Detected"] = "Detected";
    TunnelType["Forwarded"] = "Forwarded";
    TunnelType["Add"] = "Add";
})(TunnelType || (TunnelType = {})));
var TunnelEditId;
( (function(TunnelEditId) {
    TunnelEditId[TunnelEditId["None"] = 0] = "None";
    TunnelEditId[TunnelEditId["New"] = 1] = "New";
    TunnelEditId[TunnelEditId["Label"] = 2] = "Label";
    TunnelEditId[TunnelEditId["LocalPort"] = 3] = "LocalPort";
})(TunnelEditId || (TunnelEditId = {})));
export { IRemoteExplorerService, PORT_AUTO_FORWARD_SETTING, PORT_AUTO_SOURCE_SETTING, PORT_AUTO_SOURCE_SETTING_OUTPUT, TUNNEL_VIEW_ID, TunnelEditId, TunnelType };
