// vscode/vscode/src/vs/workbench/api/browser/mainThreadQuickDiff.js
import { __decorate, __param } from '../../../../../../external/tslib/tslib.es6.js';
import { CancellationTokenSource } from 'monaco-editor/esm/vs/base/common/cancellation.js';
import { dispose } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import { ExtHostContext, MainContext } from '../common/extHost.protocol.js';
import { IQuickDiffService } from '../../contrib/scm/common/quickDiff.js';
import { extHostNamedCustomer } from '../../services/extensions/common/extHostCustomers.js';
let MainThreadQuickDiff = class MainThreadQuickDiff {
    constructor(extHostContext, quickDiffService) {
        this.quickDiffService = quickDiffService;
        this.providers = ( new Map());
        this.providerDisposables = ( new Map());
        this.proxy = ( extHostContext.getProxy(ExtHostContext.ExtHostQuickDiff));
    }
    async $registerQuickDiffProvider(handle, selector, label, rootUri) {
        const provider = {
            label,
            rootUri: URI.revive(rootUri),
            selector,
            isSCM: false,
            getOriginalResource: async (uri) => {
                return URI.revive(await this.proxy.$provideOriginalResource(handle, uri, ( new CancellationTokenSource()).token));
            }
        };
        this.providers.set(handle, provider);
        const disposable = this.quickDiffService.addQuickDiffProvider(provider);
        this.providerDisposables.set(handle, disposable);
    }
    async $unregisterQuickDiffProvider(handle) {
        if (( this.providers.has(handle))) {
            this.providers.delete(handle);
        }
        if (( this.providerDisposables.has(handle))) {
            this.providerDisposables.delete(handle);
        }
    }
    dispose() {
        this.providers.clear();
        dispose(( this.providerDisposables.values()));
        this.providerDisposables.clear();
    }
};
MainThreadQuickDiff = __decorate([
    extHostNamedCustomer(MainContext.MainThreadQuickDiff),
    ( __param(1, IQuickDiffService))
], MainThreadQuickDiff);
export { MainThreadQuickDiff };
