// vscode/vscode/src/vs/workbench/api/browser/mainThreadTesting.js
import { __decorate, __param } from '../../../../../../external/tslib/tslib.es6.js';
import { Disposable, MutableDisposable, toDisposable, DisposableStore } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { revive } from 'monaco-editor/esm/vs/base/common/marshalling.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import { Range } from 'monaco-editor/esm/vs/editor/common/core/range.js';
import { MutableObservableValue } from '../../contrib/testing/common/observableValue.js';
import { ITestItem, ITestMessage, TestsDiffOp } from '../../contrib/testing/common/testTypes.js';
import { TestCoverage } from '../../contrib/testing/common/testCoverage.js';
import { ITestProfileService } from '../../contrib/testing/common/testProfileService.js';
import { LiveTestResult } from '../../contrib/testing/common/testResult.js';
import { ITestResultService } from '../../contrib/testing/common/testResultService.js';
import { ITestService } from '../../contrib/testing/common/testService.js';
import { extHostNamedCustomer } from '../../services/extensions/common/extHostCustomers.js';
import { ExtHostContext, MainContext } from '../common/extHost.protocol.js';
import { WellDefinedPrefixTree } from '../../../base/common/prefixTree.js';
import { TestId } from '../../contrib/testing/common/testId.js';
let MainThreadTesting = class MainThreadTesting extends Disposable {
    constructor(extHostContext, testService, testProfiles, resultService) {
        super();
        this.testService = testService;
        this.testProfiles = testProfiles;
        this.resultService = resultService;
        this.diffListener = this._register(( new MutableDisposable()));
        this.testProviderRegistrations = ( new Map());
        this.proxy = ( extHostContext.getProxy(ExtHostContext.ExtHostTesting));
        this._register(this.testService.onDidCancelTestRun(({ runId }) => {
            this.proxy.$cancelExtensionTestRun(runId);
        }));
        this._register(resultService.onResultsChanged(evt => {
            const results = 'completed' in evt ? evt.completed : ('inserted' in evt ? evt.inserted : undefined);
            const serialized = results?.toJSONWithMessages();
            if (serialized) {
                this.proxy.$publishTestResults([serialized]);
            }
        }));
    }
    $markTestRetired(testIds) {
        let tree;
        if (testIds) {
            tree = ( new WellDefinedPrefixTree());
            for (const id of testIds) {
                tree.insert(TestId.fromString(id).path, undefined);
            }
        }
        for (const result of this.resultService.results) {
            if (result instanceof LiveTestResult) {
                result.markRetired(tree);
            }
        }
    }
    $publishTestRunProfile(profile) {
        const controller = this.testProviderRegistrations.get(profile.controllerId);
        if (controller) {
            this.testProfiles.addProfile(controller.instance, profile);
        }
    }
    $updateTestRunConfig(controllerId, profileId, update) {
        this.testProfiles.updateProfile(controllerId, profileId, update);
    }
    $removeTestProfile(controllerId, profileId) {
        this.testProfiles.removeProfile(controllerId, profileId);
    }
    $addTestsToRun(controllerId, runId, tests) {
        this.withLiveRun(runId, r => r.addTestChainToRun(controllerId, ( tests.map(ITestItem.deserialize))));
    }
    $signalCoverageAvailable(runId, taskId) {
        this.withLiveRun(runId, run => {
            const task = run.tasks.find(t => t.id === taskId);
            if (!task) {
                return;
            }
            task.coverage.value = ( new TestCoverage({
                provideFileCoverage: async (token) => revive(await this.proxy.$provideFileCoverage(runId, taskId, token)),
                resolveFileCoverage: (i, token) => this.proxy.$resolveFileCoverage(runId, taskId, i, token),
            }));
        });
    }
    $startedExtensionTestRun(req) {
        this.resultService.createLiveResult(req);
    }
    $startedTestRunTask(runId, task) {
        this.withLiveRun(runId, r => r.addTask(task));
    }
    $finishedTestRunTask(runId, taskId) {
        this.withLiveRun(runId, r => r.markTaskComplete(taskId));
    }
    $finishedExtensionTestRun(runId) {
        this.withLiveRun(runId, r => r.markComplete());
    }
    $updateTestStateInRun(runId, taskId, testId, state, duration) {
        this.withLiveRun(runId, r => r.updateState(testId, taskId, state, duration));
    }
    $appendOutputToRun(runId, taskId, output, locationDto, testId) {
        const location = locationDto && {
            uri: URI.revive(locationDto.uri),
            range: Range.lift(locationDto.range)
        };
        this.withLiveRun(runId, r => r.appendOutput(output, taskId, location, testId));
    }
    $appendTestMessagesInRun(runId, taskId, testId, messages) {
        const r = this.resultService.getResult(runId);
        if (r && r instanceof LiveTestResult) {
            for (const message of messages) {
                r.appendMessage(testId, taskId, ITestMessage.deserialize(message));
            }
        }
    }
    $registerTestController(controllerId, labelStr, canRefreshValue) {
        const disposable = ( new DisposableStore());
        const label = disposable.add(( new MutableObservableValue(labelStr)));
        const canRefresh = disposable.add(( new MutableObservableValue(canRefreshValue)));
        const controller = {
            id: controllerId,
            label,
            canRefresh,
            syncTests: () => this.proxy.$syncTests(),
            refreshTests: token => this.proxy.$refreshTests(controllerId, token),
            configureRunProfile: id => this.proxy.$configureRunProfile(controllerId, id),
            runTests: (reqs, token) => this.proxy.$runControllerTests(reqs, token),
            startContinuousRun: (reqs, token) => this.proxy.$startContinuousRun(reqs, token),
            expandTest: (testId, levels) => this.proxy.$expandTest(testId, isFinite(levels) ? levels : -1),
        };
        disposable.add(toDisposable(() => this.testProfiles.removeProfile(controllerId)));
        disposable.add(this.testService.registerTestController(controllerId, controller));
        this.testProviderRegistrations.set(controllerId, {
            instance: controller,
            label,
            canRefresh,
            disposable
        });
    }
    $updateController(controllerId, patch) {
        const controller = this.testProviderRegistrations.get(controllerId);
        if (!controller) {
            return;
        }
        if (patch.label !== undefined) {
            controller.label.value = patch.label;
        }
        if (patch.canRefresh !== undefined) {
            controller.canRefresh.value = patch.canRefresh;
        }
    }
    $unregisterTestController(controllerId) {
        this.testProviderRegistrations.get(controllerId)?.disposable.dispose();
        this.testProviderRegistrations.delete(controllerId);
    }
    $subscribeToDiffs() {
        this.proxy.$acceptDiff(( this.testService.collection.getReviverDiff().map(TestsDiffOp.serialize)));
        this.diffListener.value = this.testService.onDidProcessDiff(this.proxy.$acceptDiff, this.proxy);
    }
    $unsubscribeFromDiffs() {
        this.diffListener.clear();
    }
    $publishDiff(controllerId, diff) {
        this.testService.publishDiff(controllerId, ( diff.map(TestsDiffOp.deserialize)));
    }
    async $runTests(req, token) {
        const result = await this.testService.runResolvedTests(req, token);
        return result.id;
    }
    dispose() {
        super.dispose();
        for (const subscription of ( this.testProviderRegistrations.values())) {
            subscription.disposable.dispose();
        }
        this.testProviderRegistrations.clear();
    }
    withLiveRun(runId, fn) {
        const r = this.resultService.getResult(runId);
        return r && r instanceof LiveTestResult ? fn(r) : undefined;
    }
};
MainThreadTesting = __decorate([
    extHostNamedCustomer(MainContext.MainThreadTesting),
    ( __param(1, ITestService)),
    ( __param(2, ITestProfileService)),
    ( __param(3, ITestResultService))
], MainThreadTesting);
export { MainThreadTesting };
