// @codingame/monaco-vscode-configuration-service-override/services.js
import './missing-services.js';
export { default as Severity } from 'monaco-editor/esm/vs/base/common/severity.js';
export { StorageScope, StorageTarget } from 'vscode/vscode/vs/platform/storage/common/storage';
import { StandaloneServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices.js';
export { StandaloneServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices.js';
export { IInstantiationService } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
import { Disposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import 'vscode/service-override/layout';
import 'vscode/service-override/environment';
import 'vscode/service-override/extensions';
import 'vscode/service-override/files';
import 'vscode/service-override/quickaccess';
import { serviceInitializedBarrier } from 'vscode/lifecycle';
import 'vscode/workbench';
export { SyncDescriptor } from 'monaco-editor/esm/vs/platform/instantiation/common/descriptors.js';
export { ICommandService } from 'monaco-editor/esm/vs/platform/commands/common/commands.js';
export { INotificationService, NeverShowAgainScope, NoOpNotification, NoOpProgress, NotificationPriority, NotificationsFilter } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
export { IBulkEditService } from 'monaco-editor/esm/vs/editor/browser/services/bulkEditService.js';
export { ICodeEditorService } from 'monaco-editor/esm/vs/editor/browser/services/codeEditorService.js';
export { ILanguageService } from 'monaco-editor/esm/vs/editor/common/languages/language.js';
export { ILanguageConfigurationService } from 'monaco-editor/esm/vs/editor/common/languages/languageConfigurationRegistry.js';
export { IEditorWorkerService } from 'monaco-editor/esm/vs/editor/common/services/editorWorker.js';
export { ILanguageFeaturesService } from 'monaco-editor/esm/vs/editor/common/services/languageFeatures.js';
export { IModelService } from 'monaco-editor/esm/vs/editor/common/services/model.js';
export { ITextModelService } from 'monaco-editor/esm/vs/editor/common/services/resolverService.js';
export { IClipboardService } from 'monaco-editor/esm/vs/platform/clipboard/common/clipboardService.js';
import './override/vs/platform/dialogs/common/dialogs.js';
export { IFileService } from 'monaco-editor/esm/vs/platform/files/common/files.js';
export { IMarkerService } from 'monaco-editor/esm/vs/platform/markers/common/markers.js';
export { IOpenerService } from 'monaco-editor/esm/vs/platform/opener/common/opener.js';
export { IProductService } from 'monaco-editor/esm/vs/platform/product/common/productService.js';
export { IQuickInputService } from 'monaco-editor/esm/vs/platform/quickinput/common/quickInput.js';
export { ITelemetryService } from 'monaco-editor/esm/vs/platform/telemetry/common/telemetry.js';
export { IUriIdentityService } from 'monaco-editor/esm/vs/platform/uriIdentity/common/uriIdentity.js';
export { IBreadcrumbsService } from 'vscode/vscode/vs/workbench/browser/parts/editor/breadcrumbs';
export { GroupDirection, GroupLocation, GroupOrientation, GroupsArrangement, GroupsOrder, IEditorGroupsService, MergeGroupMode } from 'vscode/vscode/vs/workbench/services/editor/common/editorGroupsService';
export { CloseDirection, EditorsOrder } from 'vscode/vscode/vs/workbench/common/editor';
export { EditorInput } from 'vscode/vscode/vs/workbench/common/editor/editorInput';
export { IEditorService } from 'vscode/vscode/vs/workbench/services/editor/common/editorService';
export { IEditorResolverService } from 'vscode/vscode/vs/workbench/services/editor/common/editorResolverService';
export { ITextEditorService } from 'vscode/vscode/vs/workbench/services/textfile/common/textEditorService';
export { IWorkbenchEnvironmentService } from 'vscode/vscode/vs/workbench/services/environment/common/environmentService';
export { IHostService } from 'vscode/vscode/vs/workbench/services/host/browser/host';
export { ILanguageStatusService } from 'vscode/vscode/vs/workbench/services/languageStatus/common/languageStatusService';
export { IPaneCompositePartService } from 'vscode/vscode/vs/workbench/services/panecomposite/browser/panecomposite';
export { IPathService } from 'vscode/vscode/vs/workbench/services/path/common/pathService';
export { ITextFileService } from 'vscode/vscode/vs/workbench/services/textfile/common/textfiles';
export { IWorkingCopyFileService } from 'vscode/vscode/vs/workbench/services/workingCopy/common/workingCopyFileService';
export { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
export { IContextKeyService } from 'monaco-editor/esm/vs/platform/contextkey/common/contextkey.js';
export { IThemeService } from 'monaco-editor/esm/vs/platform/theme/common/themeService.js';
export { ISnippetsService } from 'vscode/vscode/vs/workbench/contrib/snippets/browser/snippets';
export { IWorkspaceContextService } from 'monaco-editor/esm/vs/platform/workspace/common/workspace.js';
export { IStorageService } from 'monaco-editor/esm/vs/platform/storage/common/storage.js';
export { IAudioCueService } from 'monaco-editor/esm/vs/platform/audioCues/browser/audioCueService.js';
export { IDebugService } from 'vscode/vscode/vs/workbench/contrib/debug/common/debug';
export { ILogService, ILoggerService, LogLevel } from 'monaco-editor/esm/vs/platform/log/common/log.js';
export { IViewDescriptorService, IViewsService } from 'vscode/vscode/vs/workbench/common/views';
export { ILayoutService } from 'monaco-editor/esm/vs/platform/layout/browser/layoutService.js';
export { IPreferencesService } from 'vscode/vscode/vs/workbench/services/preferences/common/preferences';
export { IPreferencesSearchService } from 'vscode/vscode/vs/workbench/contrib/preferences/common/preferences';
export { IKeybindingEditingService } from 'vscode/vscode/vs/workbench/services/keybinding/common/keybindingEditing';
export { IOutputService } from 'vscode/vscode/vs/workbench/services/output/common/output';
export { GoFilter, GoScope, IHistoryService } from 'vscode/vscode/vs/workbench/services/history/common/history';
export { IRemoteAgentService } from 'vscode/vscode/vs/workbench/services/remote/common/remoteAgentService';
export { IRemoteAuthorityResolverService } from 'vscode/vscode/vs/platform/remote/common/remoteAuthorityResolver';
export { IRemoteSocketFactoryService } from 'vscode/vscode/vs/platform/remote/common/remoteSocketFactoryService';
export { ITerminalInstanceService, ITerminalService } from 'vscode/vscode/vs/workbench/contrib/terminal/browser/terminal';
export { IFilesConfigurationService } from 'vscode/vscode/vs/workbench/services/filesConfiguration/common/filesConfigurationService';
export { ILabelService } from 'monaco-editor/esm/vs/platform/label/common/label.js';
export { IActivityService } from 'vscode/vscode/vs/workbench/services/activity/common/activity';
export { IHoverService } from 'vscode/vscode/vs/workbench/services/hover/browser/hover';
export { IExplorerService } from 'vscode/vscode/vs/workbench/contrib/files/browser/files';
export { IStatusbarService } from 'vscode/vscode/vs/workbench/services/statusbar/browser/statusbar';
export { ITitleService } from 'vscode/vscode/vs/workbench/services/title/common/titleService';
export { IBannerService } from 'vscode/vscode/vs/workbench/services/banner/browser/bannerService';
export { IWorkspaceTrustEnablementService, IWorkspaceTrustManagementService, IWorkspaceTrustRequestService } from 'monaco-editor/esm/vs/platform/workspace/common/workspaceTrust.js';
export { IDialogService, IFileDialogService } from 'vscode/vscode/vs/platform/dialogs/common/dialogs';

StandaloneServices.withServices(() => {
    return Disposable.None;
});
async function waitServicesReady() {
    await serviceInitializedBarrier.wait();
}
async function getService(identifier) {
    await waitServicesReady();
    return StandaloneServices.get(identifier);
}

export { getService, waitServicesReady };
