// vscode/vscode/src/vs/workbench/services/workingCopy/common/workingCopyFileOperationParticipant.js
import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { ILogService } from 'monaco-editor/esm/vs/platform/log/common/log.js';
import { Disposable, toDisposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
import { LinkedList } from 'monaco-editor/esm/vs/base/common/linkedList.js';
let WorkingCopyFileOperationParticipant = class WorkingCopyFileOperationParticipant extends Disposable {
    constructor(logService, configurationService) {
        super();
        this.logService = logService;
        this.configurationService = configurationService;
        this.participants = ( new LinkedList());
    }
    addFileOperationParticipant(participant) {
        const remove = this.participants.push(participant);
        return toDisposable(() => remove());
    }
    async participate(files, operation, undoInfo, token) {
        const timeout = this.configurationService.getValue('files.participants.timeout');
        if (typeof timeout !== 'number' || timeout <= 0) {
            return;
        }
        for (const participant of this.participants) {
            try {
                await participant.participate(files, operation, undoInfo, timeout, token);
            }
            catch (err) {
                this.logService.warn(err);
            }
        }
    }
    dispose() {
        this.participants.clear();
        super.dispose();
    }
};
WorkingCopyFileOperationParticipant = ( __decorate([
    ( __param(0, ILogService)),
    ( __param(1, IConfigurationService))
], WorkingCopyFileOperationParticipant));
export { WorkingCopyFileOperationParticipant };
