// vscode/vscode/src/vs/base/browser/broadcast.js
import { getErrorMessage } from 'monaco-editor/esm/vs/base/common/errors.js';
import { Emitter } from 'monaco-editor/esm/vs/base/common/event.js';
import { Disposable, toDisposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
class BroadcastDataChannel extends Disposable {
    constructor(channelName) {
        super();
        this.channelName = channelName;
        this._onDidReceiveData = this._register(( new Emitter()));
        this.onDidReceiveData = this._onDidReceiveData.event;
        if ('BroadcastChannel' in window) {
            try {
                this.broadcastChannel = ( new BroadcastChannel(channelName));
                const listener = (event) => {
                    this._onDidReceiveData.fire(event.data);
                };
                this.broadcastChannel.addEventListener('message', listener);
                this._register(toDisposable(() => {
                    if (this.broadcastChannel) {
                        this.broadcastChannel.removeEventListener('message', listener);
                        this.broadcastChannel.close();
                    }
                }));
            }
            catch (error) {
                console.warn('Error while creating broadcast channel. Falling back to localStorage.', getErrorMessage(error));
            }
        }
        if (!this.broadcastChannel) {
            this.channelName = `BroadcastDataChannel.${channelName}`;
            this.createBroadcastChannel();
        }
    }
    createBroadcastChannel() {
        const listener = (event) => {
            if (event.key === this.channelName && event.newValue) {
                this._onDidReceiveData.fire(JSON.parse(event.newValue));
            }
        };
        window.addEventListener('storage', listener);
        this._register(toDisposable(() => window.removeEventListener('storage', listener)));
    }
    postData(data) {
        if (this.broadcastChannel) {
            this.broadcastChannel.postMessage(data);
        }
        else {
            window.localStorage.removeItem(this.channelName);
            window.localStorage.setItem(this.channelName, JSON.stringify(data));
        }
    }
}
export { BroadcastDataChannel };
