// vscode/vscode/src/vs/workbench/api/common/extHost.api.impl.js
import { CancellationTokenSource } from 'monaco-editor/esm/vs/base/common/cancellation.js';
import * as errors from 'monaco-editor/esm/vs/base/common/errors.js';
import { Emitter, Event } from 'monaco-editor/esm/vs/base/common/event.js';
import Severity from 'monaco-editor/esm/vs/base/common/severity.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import { TextEditorCursorStyle } from 'monaco-editor/esm/vs/editor/common/config/editorOptions.js';
import { OverviewRulerLane } from 'monaco-editor/esm/vs/editor/common/model.js';
import * as languageConfiguration from 'monaco-editor/esm/vs/editor/common/languages/languageConfiguration.js';
import { score } from 'monaco-editor/esm/vs/editor/common/languageSelector.js';
import * as files from 'monaco-editor/esm/vs/platform/files/common/files.js';
import { ExtHostContext, MainContext, CandidatePortSource } from './extHost.protocol.js';
import { UIKind } from '../../services/extensions/common/extensionHostProtocol.js';
import { ExtHostApiCommands } from './extHostApiCommands.js';
import { ExtHostClipboard } from './extHostClipboard.js';
import { IExtHostCommands } from './extHostCommands.js';
import { createExtHostComments } from './extHostComments.js';
import { IExtHostConfiguration } from './extHostConfiguration.js';
import { ExtHostDiagnostics } from './extHostDiagnostics.js';
import { ExtHostDialogs } from './extHostDialogs.js';
import { ExtHostDocumentContentProvider } from './extHostDocumentContentProviders.js';
import { ExtHostDocumentSaveParticipant } from './extHostDocumentSaveParticipant.js';
import { ExtHostDocuments } from './extHostDocuments.js';
import { IExtHostDocumentsAndEditors } from './extHostDocumentsAndEditors.js';
import { IExtHostExtensionService, Extension } from './extHostExtensionService.js';
import { ExtHostFileSystem } from './extHostFileSystem.js';
import { ExtHostFileSystemEventService } from './extHostFileSystemEventService.js';
import { ExtHostLanguageFeatures } from './extHostLanguageFeatures.js';
import { ExtHostLanguages } from './extHostLanguages.js';
import { ExtHostMessageService } from './extHostMessageService.js';
import { IExtHostOutputService } from './extHostOutput.js';
import { ExtHostProgress } from './extHostProgress.js';
import { createExtHostQuickOpen } from './extHostQuickOpen.js';
import { ExtHostSCM } from './extHostSCM.js';
import { ExtHostStatusBar } from './extHostStatusBar.js';
import { IExtHostStorage } from './extHostStorage.js';
import { IExtHostTerminalService } from './extHostTerminalService.js';
import { ExtHostEditors } from './extHostTextEditors.js';
import { ExtHostTreeViews } from './extHostTreeViews.js';
import { LanguageSelector } from './extHostTypeConverters.js';
import { Breakpoint, TerminalOutputAnchor, ChatMessage, ChatMessageRole, ChatVariableLevel, CallHierarchyIncomingCall, CallHierarchyItem, CallHierarchyOutgoingCall, CodeAction, CodeActionKind, CodeActionTriggerKind, CodeLens, Color, ColorInformation, ColorPresentation, ColorThemeKind, CommentMode, CommentState, CommentThreadCollapsibleState, CommentThreadState, CompletionItem, CompletionItemKind, CompletionItemTag, CompletionList, CompletionTriggerKind, ConfigurationTarget, CustomExecution, DebugAdapterExecutable, DebugAdapterInlineImplementation, DebugAdapterNamedPipeServer, DebugAdapterServer, DebugConsoleMode, DecorationRangeBehavior, Diagnostic, DiagnosticRelatedInformation, DiagnosticSeverity, DiagnosticTag, Disposable, DocumentHighlight, DocumentHighlightKind, DocumentLink, DocumentSymbol, EndOfLine, EnvironmentVariableMutatorType, EvaluatableExpression, InlineValueText, InlineValueVariableLookup, InlineValueEvaluatableExpression, InlineCompletionTriggerKind, ExtensionKind, ExtensionMode, ExternalUriOpenerPriority, FileChangeType, FileDecoration, FileSystemError, FoldingRange, FoldingRangeKind, FunctionBreakpoint, InlineSuggestion, InlineSuggestionList, Hover, Location, MarkdownString, ParameterInformation, PortAutoForwardAction, Position, ProcessExecution, ProgressLocation, QuickInputButtons, Range, RelativePattern, Selection, SelectionRange, SemanticTokens, SemanticTokensBuilder, SemanticTokensEdit, SemanticTokensEdits, SemanticTokensLegend, ShellExecution, ShellQuoting, SignatureHelp, SignatureHelpTriggerKind, SignatureInformation, SnippetString, SourceBreakpoint, StandardTokenType, StatusBarAlignment, SymbolInformation, SymbolKind, SymbolTag, Task, TaskGroup, TaskPanelKind, TaskRevealKind, TaskScope, TerminalLink, TerminalQuickFixCommand, TerminalQuickFixOpener, TerminalLocation, TerminalProfile, TerminalExitReason, TextDocumentSaveReason, TextEdit, SnippetTextEdit, TextEditorLineNumbersStyle, TextEditorRevealType, TextEditorSelectionChangeKind, SyntaxTokenType, TextDocumentChangeReason, ThemeColor, ThemeIcon, TreeItem, TreeItemCheckboxState, TreeItemCollapsibleState, TypeHierarchyItem, ViewColumn, WorkspaceEdit, DocumentDropEdit, DocumentPasteEdit, InlayHint, InlayHintLabelPart, InlayHintKind, RemoteAuthorityResolverError, ResolvedAuthority, ManagedResolvedAuthority, SourceControlInputBoxValidationType, ExtensionRuntime, TimelineItem, NotebookRange, NotebookCellKind, NotebookCellExecutionState, NotebookCellData, NotebookData, NotebookRendererScript, NotebookCellStatusBarAlignment, NotebookEditorRevealType, NotebookCellOutput, NotebookCellOutputItem, NotebookCellStatusBarItem, NotebookControllerAffinity, NotebookControllerAffinity2, NotebookEdit, NotebookKernelSourceAction, PortAttributes, LinkedEditingRanges, TestResultState, TestRunRequest, TestMessage, TestTag, TestRunProfileKind, DataTransfer, DataTransferItem, CoveredCount, FileCoverage, StatementCoverage, BranchCoverage, FunctionCoverage, WorkspaceTrustState, LanguageStatusSeverity, QuickPickItemKind, InputBoxValidationSeverity, TextTabInput, TextDiffTabInput, TextMergeTabInput, CustomEditorTabInput, NotebookEditorTabInput, NotebookDiffEditorTabInput, WebviewEditorTabInput, TerminalEditorTabInput, InteractiveWindowInput, InteractiveSessionVoteDirection, InteractiveSessionCopyKind, InteractiveEditorResponseFeedbackKind, StackFrameFocus, ThreadFocus, RelatedInformationType } from './extHostTypes.js';
import { TelemetryTrustedValue } from '../../../platform/telemetry/common/telemetryUtils.js';
import { ExtHostUrls } from './extHostUrls.js';
import { ExtHostWebviews } from './extHostWebview.js';
import { IExtHostWindow } from './extHostWindow.js';
import { IExtHostWorkspace } from './extHostWorkspace.js';
import { ExtensionIdentifierSet } from 'monaco-editor/esm/vs/platform/extensions/common/extensions.js';
import { ExtHostEditorInsets } from './extHostCodeInsets.js';
import { ExtHostLabelService } from './extHostLabelService.js';
import { getRemoteName } from 'monaco-editor/esm/vs/platform/remote/common/remoteHosts.js';
import { IExtHostDecorations } from './extHostDecorations.js';
import { IExtHostTask } from './extHostTask.js';
import { IExtHostDebugService } from './extHostDebugService.js';
import { IExtHostSearch } from './extHostSearch.js';
import { ILoggerService, ILogService, LogLevel } from 'monaco-editor/esm/vs/platform/log/common/log.js';
import { IURITransformerService } from './extHostUriTransformerService.js';
import { IExtHostRpcService } from './extHostRpcService.js';
import { IExtHostInitDataService } from './extHostInitDataService.js';
import { ExtHostNotebookController } from './extHostNotebook.js';
import { ExtHostTheming } from './extHostTheming.js';
import { IExtHostTunnelService } from './extHostTunnelService.js';
import { IExtHostApiDeprecationService } from './extHostApiDeprecationService.js';
import { ExtHostAuthentication } from './extHostAuthentication.js';
import { ExtHostTimeline } from './extHostTimeline.js';
import { IExtensionStoragePaths } from './extHostStoragePaths.js';
import { IExtHostConsumerFileSystem } from './extHostFileSystemConsumer.js';
import { ExtHostWebviewViews } from './extHostWebviewView.js';
import { ExtHostCustomEditors } from './extHostCustomEditors.js';
import { ExtHostWebviewPanels } from './extHostWebviewPanels.js';
import { ExtHostBulkEdits } from './extHostBulkEdits.js';
import { IExtHostFileSystemInfo } from './extHostFileSystemInfo.js';
import { ExtHostTesting } from './extHostTesting.js';
import { ExtHostUriOpeners } from './extHostUriOpener.js';
import { IExtHostSecretState } from './extHostSecretState.js';
import { IExtHostEditorTabs } from './extHostEditorTabs.js';
import { IExtHostTelemetry, isNewAppInstall, ExtHostTelemetryLogger } from './extHostTelemetry.js';
import { ExtHostNotebookKernels } from './extHostNotebookKernels.js';
import { TextSearchCompleteMessageType } from '../../services/search/common/searchExtTypes.js';
import { ExtHostNotebookRenderers } from './extHostNotebookRenderers.js';
import { Schemas } from 'monaco-editor/esm/vs/base/common/network.js';
import { matchesScheme } from 'monaco-editor/esm/vs/platform/opener/common/opener.js';
import { ExtHostNotebookEditors } from './extHostNotebookEditors.js';
import { ExtHostNotebookDocuments } from './extHostNotebookDocuments.js';
import { ExtHostInteractive } from './extHostInteractive.js';
import { combinedDisposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { checkProposedApiEnabled, isProposedApiEnabled } from '../../services/extensions/common/extensions.js';
import { DebugConfigurationProviderTriggerKind } from '../../contrib/debug/common/debug.js';
import { IExtHostLocalizationService } from './extHostLocalizationService.js';
import { EditSessionIdentityMatch } from '../../../platform/workspace/common/editSessions.js';
import { ExtHostProfileContentHandlers } from './extHostProfileContentHandler.js';
import { ExtHostQuickDiff } from './extHostQuickDiff.js';
import { ExtHostChat } from './extHostChat.js';
import { ExtHostInteractiveEditor } from './extHostInlineChat.js';
import { ExtHostNotebookDocumentSaveParticipant } from './extHostNotebookDocumentSaveParticipant.js';
import { ExtHostIssueReporter } from './extHostIssueReporter.js';
import { IExtHostManagedSockets } from './extHostManagedSockets.js';
import { ExtHostShare } from './extHostShare.js';
import { ExtHostChatProvider } from './extHostChatProvider.js';
import { ExtHostChatSlashCommands } from './extHostChatSlashCommand.js';
import { ExtHostChatVariables } from './extHostChatVariables.js';
import { ExtHostRelatedInformation } from './extHostAiRelatedInformation.js';
import { ExtHostAiEmbeddingVector } from './extHostEmbeddingVector.js';
import { ExtHostChatAgents } from './extHostChatAgents.js';
function createApiFactoryAndRegisterActors(accessor) {
    const initData = accessor.get(IExtHostInitDataService);
    const extHostFileSystemInfo = accessor.get(IExtHostFileSystemInfo);
    const extHostConsumerFileSystem = accessor.get(IExtHostConsumerFileSystem);
    const extensionService = accessor.get(IExtHostExtensionService);
    const extHostWorkspace = accessor.get(IExtHostWorkspace);
    const extHostTelemetry = accessor.get(IExtHostTelemetry);
    const extHostConfiguration = accessor.get(IExtHostConfiguration);
    const uriTransformer = accessor.get(IURITransformerService);
    const rpcProtocol = accessor.get(IExtHostRpcService);
    const extHostStorage = accessor.get(IExtHostStorage);
    const extensionStoragePaths = accessor.get(IExtensionStoragePaths);
    const extHostLoggerService = accessor.get(ILoggerService);
    const extHostLogService = accessor.get(ILogService);
    const extHostTunnelService = accessor.get(IExtHostTunnelService);
    const extHostApiDeprecation = accessor.get(IExtHostApiDeprecationService);
    const extHostWindow = accessor.get(IExtHostWindow);
    const extHostSecretState = accessor.get(IExtHostSecretState);
    const extHostEditorTabs = accessor.get(IExtHostEditorTabs);
    const extHostManagedSockets = accessor.get(IExtHostManagedSockets);
    rpcProtocol.set(ExtHostContext.ExtHostFileSystemInfo, extHostFileSystemInfo);
    rpcProtocol.set(ExtHostContext.ExtHostLogLevelServiceShape, extHostLoggerService);
    rpcProtocol.set(ExtHostContext.ExtHostWorkspace, extHostWorkspace);
    rpcProtocol.set(ExtHostContext.ExtHostConfiguration, extHostConfiguration);
    rpcProtocol.set(ExtHostContext.ExtHostExtensionService, extensionService);
    rpcProtocol.set(ExtHostContext.ExtHostStorage, extHostStorage);
    rpcProtocol.set(ExtHostContext.ExtHostTunnelService, extHostTunnelService);
    rpcProtocol.set(ExtHostContext.ExtHostWindow, extHostWindow);
    rpcProtocol.set(ExtHostContext.ExtHostSecretState, extHostSecretState);
    rpcProtocol.set(ExtHostContext.ExtHostTelemetry, extHostTelemetry);
    rpcProtocol.set(ExtHostContext.ExtHostEditorTabs, extHostEditorTabs);
    rpcProtocol.set(ExtHostContext.ExtHostManagedSockets, extHostManagedSockets);
    const extHostDecorations = rpcProtocol.set(ExtHostContext.ExtHostDecorations, accessor.get(IExtHostDecorations));
    const extHostDocumentsAndEditors = rpcProtocol.set(ExtHostContext.ExtHostDocumentsAndEditors, accessor.get(IExtHostDocumentsAndEditors));
    const extHostCommands = rpcProtocol.set(ExtHostContext.ExtHostCommands, accessor.get(IExtHostCommands));
    const extHostTerminalService = rpcProtocol.set(ExtHostContext.ExtHostTerminalService, accessor.get(IExtHostTerminalService));
    const extHostDebugService = rpcProtocol.set(ExtHostContext.ExtHostDebugService, accessor.get(IExtHostDebugService));
    const extHostSearch = rpcProtocol.set(ExtHostContext.ExtHostSearch, accessor.get(IExtHostSearch));
    const extHostTask = rpcProtocol.set(ExtHostContext.ExtHostTask, accessor.get(IExtHostTask));
    const extHostOutputService = rpcProtocol.set(ExtHostContext.ExtHostOutputService, accessor.get(IExtHostOutputService));
    const extHostLocalization = rpcProtocol.set(ExtHostContext.ExtHostLocalization, accessor.get(IExtHostLocalizationService));
    const extHostUrls = rpcProtocol.set(ExtHostContext.ExtHostUrls, ( new ExtHostUrls(rpcProtocol)));
    const extHostDocuments = rpcProtocol.set(ExtHostContext.ExtHostDocuments, ( new ExtHostDocuments(rpcProtocol, extHostDocumentsAndEditors)));
    const extHostDocumentContentProviders = rpcProtocol.set(ExtHostContext.ExtHostDocumentContentProviders, ( new ExtHostDocumentContentProvider(rpcProtocol, extHostDocumentsAndEditors, extHostLogService)));
    const extHostDocumentSaveParticipant = rpcProtocol.set(ExtHostContext.ExtHostDocumentSaveParticipant, ( new ExtHostDocumentSaveParticipant(extHostLogService, extHostDocuments, ( rpcProtocol.getProxy(MainContext.MainThreadBulkEdits)))));
    const extHostNotebook = rpcProtocol.set(ExtHostContext.ExtHostNotebook, ( new ExtHostNotebookController(
        rpcProtocol,
        extHostCommands,
        extHostDocumentsAndEditors,
        extHostDocuments,
        extHostConsumerFileSystem
    )));
    const extHostNotebookDocuments = rpcProtocol.set(ExtHostContext.ExtHostNotebookDocuments, ( new ExtHostNotebookDocuments(extHostNotebook)));
    const extHostNotebookEditors = rpcProtocol.set(ExtHostContext.ExtHostNotebookEditors, ( new ExtHostNotebookEditors(extHostLogService, extHostNotebook)));
    const extHostNotebookKernels = rpcProtocol.set(ExtHostContext.ExtHostNotebookKernels, ( new ExtHostNotebookKernels(rpcProtocol, initData, extHostNotebook, extHostCommands, extHostLogService)));
    const extHostNotebookRenderers = rpcProtocol.set(ExtHostContext.ExtHostNotebookRenderers, ( new ExtHostNotebookRenderers(rpcProtocol, extHostNotebook)));
    const extHostNotebookDocumentSaveParticipant = rpcProtocol.set(ExtHostContext.ExtHostNotebookDocumentSaveParticipant, ( new ExtHostNotebookDocumentSaveParticipant(extHostLogService, extHostNotebook, ( rpcProtocol.getProxy(MainContext.MainThreadBulkEdits)))));
    const extHostEditors = rpcProtocol.set(ExtHostContext.ExtHostEditors, ( new ExtHostEditors(rpcProtocol, extHostDocumentsAndEditors)));
    const extHostTreeViews = rpcProtocol.set(ExtHostContext.ExtHostTreeViews, ( new ExtHostTreeViews(( rpcProtocol.getProxy(MainContext.MainThreadTreeViews)), extHostCommands, extHostLogService)));
    const extHostEditorInsets = rpcProtocol.set(ExtHostContext.ExtHostEditorInsets, ( new ExtHostEditorInsets(( rpcProtocol.getProxy(MainContext.MainThreadEditorInsets)), extHostEditors, initData.remote)));
    const extHostDiagnostics = rpcProtocol.set(ExtHostContext.ExtHostDiagnostics, ( new ExtHostDiagnostics(
        rpcProtocol,
        extHostLogService,
        extHostFileSystemInfo,
        extHostDocumentsAndEditors
    )));
    const extHostLanguages = rpcProtocol.set(ExtHostContext.ExtHostLanguages, ( new ExtHostLanguages(rpcProtocol, extHostDocuments, extHostCommands.converter, uriTransformer)));
    const extHostLanguageFeatures = rpcProtocol.set(ExtHostContext.ExtHostLanguageFeatures, ( new ExtHostLanguageFeatures(
        rpcProtocol,
        uriTransformer,
        extHostDocuments,
        extHostCommands,
        extHostDiagnostics,
        extHostLogService,
        extHostApiDeprecation,
        extHostTelemetry
    )));
    const extHostFileSystem = rpcProtocol.set(ExtHostContext.ExtHostFileSystem, ( new ExtHostFileSystem(rpcProtocol, extHostLanguageFeatures)));
    const extHostFileSystemEvent = rpcProtocol.set(ExtHostContext.ExtHostFileSystemEventService, ( new ExtHostFileSystemEventService(rpcProtocol, extHostLogService, extHostDocumentsAndEditors)));
    const extHostQuickOpen = rpcProtocol.set(ExtHostContext.ExtHostQuickOpen, createExtHostQuickOpen(rpcProtocol, extHostWorkspace, extHostCommands));
    const extHostSCM = rpcProtocol.set(ExtHostContext.ExtHostSCM, ( new ExtHostSCM(rpcProtocol, extHostCommands, extHostDocuments, extHostLogService)));
    const extHostQuickDiff = rpcProtocol.set(ExtHostContext.ExtHostQuickDiff, ( new ExtHostQuickDiff(rpcProtocol, uriTransformer)));
    const extHostShare = rpcProtocol.set(ExtHostContext.ExtHostShare, ( new ExtHostShare(rpcProtocol, uriTransformer)));
    const extHostComment = rpcProtocol.set(ExtHostContext.ExtHostComments, createExtHostComments(rpcProtocol, extHostCommands, extHostDocuments));
    const extHostProgress = rpcProtocol.set(ExtHostContext.ExtHostProgress, ( new ExtHostProgress(( rpcProtocol.getProxy(MainContext.MainThreadProgress)))));
    const extHostLabelService = rpcProtocol.set(ExtHostContext.ExtHosLabelService, ( new ExtHostLabelService(rpcProtocol)));
    const extHostTheming = rpcProtocol.set(ExtHostContext.ExtHostTheming, ( new ExtHostTheming(rpcProtocol)));
    const extHostAuthentication = rpcProtocol.set(ExtHostContext.ExtHostAuthentication, ( new ExtHostAuthentication(rpcProtocol)));
    const extHostTimeline = rpcProtocol.set(ExtHostContext.ExtHostTimeline, ( new ExtHostTimeline(rpcProtocol, extHostCommands)));
    const extHostWebviews = rpcProtocol.set(ExtHostContext.ExtHostWebviews, ( new ExtHostWebviews(
        rpcProtocol,
        initData.remote,
        extHostWorkspace,
        extHostLogService,
        extHostApiDeprecation
    )));
    const extHostWebviewPanels = rpcProtocol.set(ExtHostContext.ExtHostWebviewPanels, ( new ExtHostWebviewPanels(rpcProtocol, extHostWebviews, extHostWorkspace)));
    const extHostCustomEditors = rpcProtocol.set(ExtHostContext.ExtHostCustomEditors, ( new ExtHostCustomEditors(
        rpcProtocol,
        extHostDocuments,
        extensionStoragePaths,
        extHostWebviews,
        extHostWebviewPanels
    )));
    const extHostWebviewViews = rpcProtocol.set(ExtHostContext.ExtHostWebviewViews, ( new ExtHostWebviewViews(rpcProtocol, extHostWebviews)));
    const extHostTesting = rpcProtocol.set(ExtHostContext.ExtHostTesting, ( new ExtHostTesting(rpcProtocol, extHostCommands, extHostDocumentsAndEditors)));
    const extHostUriOpeners = rpcProtocol.set(ExtHostContext.ExtHostUriOpeners, ( new ExtHostUriOpeners(rpcProtocol)));
    const extHostProfileContentHandlers = rpcProtocol.set(ExtHostContext.ExtHostProfileContentHandlers, ( new ExtHostProfileContentHandlers(rpcProtocol)));
    rpcProtocol.set(ExtHostContext.ExtHostInteractive, ( new ExtHostInteractive(
        rpcProtocol,
        extHostNotebook,
        extHostDocumentsAndEditors,
        extHostCommands,
        extHostLogService
    )));
    const extHostInteractiveEditor = rpcProtocol.set(ExtHostContext.ExtHostInlineChat, ( new ExtHostInteractiveEditor(rpcProtocol, extHostCommands, extHostDocuments, extHostLogService)));
    const extHostChatProvider = rpcProtocol.set(ExtHostContext.ExtHostChatProvider, ( new ExtHostChatProvider(rpcProtocol, extHostLogService)));
    const extHostChatSlashCommands = rpcProtocol.set(ExtHostContext.ExtHostChatSlashCommands, ( new ExtHostChatSlashCommands(rpcProtocol, extHostChatProvider, extHostLogService)));
    const extHostChatAgents = rpcProtocol.set(ExtHostContext.ExtHostChatAgents, ( new ExtHostChatAgents(rpcProtocol, extHostChatProvider, extHostLogService)));
    const extHostChatVariables = rpcProtocol.set(ExtHostContext.ExtHostChatVariables, ( new ExtHostChatVariables(rpcProtocol)));
    const extHostChat = rpcProtocol.set(ExtHostContext.ExtHostChat, ( new ExtHostChat(rpcProtocol, extHostLogService)));
    const extHostAiRelatedInformation = rpcProtocol.set(ExtHostContext.ExtHostAiRelatedInformation, ( new ExtHostRelatedInformation(rpcProtocol)));
    const extHostAiEmbeddingVector = rpcProtocol.set(ExtHostContext.ExtHostAiEmbeddingVector, ( new ExtHostAiEmbeddingVector(rpcProtocol)));
    const extHostIssueReporter = rpcProtocol.set(ExtHostContext.ExtHostIssueReporter, ( new ExtHostIssueReporter(rpcProtocol)));
    const extHostStatusBar = rpcProtocol.set(ExtHostContext.ExtHostStatusBar, ( new ExtHostStatusBar(rpcProtocol, extHostCommands.converter)));
    const extHostBulkEdits = ( new ExtHostBulkEdits(rpcProtocol, extHostDocumentsAndEditors));
    const extHostClipboard = ( new ExtHostClipboard(rpcProtocol));
    const extHostMessageService = ( new ExtHostMessageService(rpcProtocol, extHostLogService));
    const extHostDialogs = ( new ExtHostDialogs(rpcProtocol));
    ExtHostApiCommands.register(extHostCommands);
    return function (extension, extensionInfo, configProvider) {
        const checkSelector = ( (function() {
            let done = !extension.isUnderDevelopment;
            function informOnce() {
                if (!done) {
                    extHostLogService.info(`Extension '${extension.identifier.value}' uses a document selector without scheme. Learn more about this: https://go.microsoft.com/fwlink/?linkid=872305`);
                    done = true;
                }
            }
            return function perform(selector) {
                if (Array.isArray(selector)) {
                    selector.forEach(perform);
                }
                else if (typeof selector === 'string') {
                    informOnce();
                }
                else {
                    const filter = selector;
                    if (typeof filter.scheme === 'undefined') {
                        informOnce();
                    }
                    if (typeof filter.exclusive === 'boolean') {
                        checkProposedApiEnabled(extension, 'documentFiltersExclusive');
                    }
                }
                return selector;
            };
        })());
        const authentication = {
            getSession(providerId, scopes, options) {
                return extHostAuthentication.getSession(extension, providerId, scopes, options);
            },
            getSessions(providerId, scopes) {
                checkProposedApiEnabled(extension, 'authGetSessions');
                return extHostAuthentication.getSessions(extension, providerId, scopes);
            },
            async hasSession(providerId, scopes) {
                checkProposedApiEnabled(extension, 'authSession');
                return !!(await extHostAuthentication.getSession(extension, providerId, scopes, { silent: true }));
            },
            get onDidChangeSessions() {
                return extHostAuthentication.onDidChangeSessions;
            },
            registerAuthenticationProvider(id, label, provider, options) {
                return extHostAuthentication.registerAuthenticationProvider(id, label, provider, options);
            }
        };
        const commands = {
            registerCommand(id, command, thisArgs) {
                return extHostCommands.registerCommand(true, id, command, thisArgs, undefined, extension);
            },
            registerTextEditorCommand(id, callback, thisArg) {
                return extHostCommands.registerCommand(true, id, (...args) => {
                    const activeTextEditor = extHostEditors.getActiveTextEditor();
                    if (!activeTextEditor) {
                        extHostLogService.warn('Cannot execute ' + id + ' because there is no active text editor.');
                        return undefined;
                    }
                    return activeTextEditor.edit((edit) => {
                        callback.apply(thisArg, [activeTextEditor, edit, ...args]);
                    }).then((result) => {
                        if (!result) {
                            extHostLogService.warn('Edits from command ' + id + ' were not applied.');
                        }
                    }, (err) => {
                        extHostLogService.warn('An error occurred while running command ' + id, err);
                    });
                }, undefined, undefined, extension);
            },
            registerDiffInformationCommand: (id, callback, thisArg) => {
                checkProposedApiEnabled(extension, 'diffCommand');
                return extHostCommands.registerCommand(true, id, async (...args) => {
                    const activeTextEditor = extHostDocumentsAndEditors.activeEditor(true);
                    if (!activeTextEditor) {
                        extHostLogService.warn('Cannot execute ' + id + ' because there is no active text editor.');
                        return undefined;
                    }
                    const diff = await extHostEditors.getDiffInformation(activeTextEditor.id);
                    callback.apply(thisArg, [diff, ...args]);
                }, undefined, undefined, extension);
            },
            executeCommand(id, ...args) {
                return extHostCommands.executeCommand(id, ...args);
            },
            getCommands(filterInternal = false) {
                return extHostCommands.getCommands(filterInternal);
            }
        };
        const env = {
            get machineId() { return initData.telemetryInfo.machineId; },
            get sessionId() { return initData.telemetryInfo.sessionId; },
            get language() { return initData.environment.appLanguage; },
            get appName() { return initData.environment.appName; },
            get appRoot() { return initData.environment.appRoot?.fsPath ?? ''; },
            get appHost() { return initData.environment.appHost; },
            get uriScheme() { return initData.environment.appUriScheme; },
            get clipboard() { return extHostClipboard.value; },
            get shell() {
                return extHostTerminalService.getDefaultShell(false);
            },
            get onDidChangeShell() {
                return extHostTerminalService.onDidChangeShell;
            },
            get isTelemetryEnabled() {
                return extHostTelemetry.getTelemetryConfiguration();
            },
            get onDidChangeTelemetryEnabled() {
                return extHostTelemetry.onDidChangeTelemetryEnabled;
            },
            get telemetryConfiguration() {
                checkProposedApiEnabled(extension, 'telemetry');
                return extHostTelemetry.getTelemetryDetails();
            },
            get onDidChangeTelemetryConfiguration() {
                checkProposedApiEnabled(extension, 'telemetry');
                return extHostTelemetry.onDidChangeTelemetryConfiguration;
            },
            get isNewAppInstall() {
                return isNewAppInstall(initData.telemetryInfo.firstSessionDate);
            },
            createTelemetryLogger(sender, options) {
                ExtHostTelemetryLogger.validateSender(sender);
                return extHostTelemetry.instantiateLogger(extension, sender, options);
            },
            openExternal(uri, options) {
                return extHostWindow.openUri(uri, {
                    allowTunneling: !!initData.remote.authority,
                    allowContributedOpeners: options?.allowContributedOpeners,
                });
            },
            async asExternalUri(uri) {
                if (uri.scheme === initData.environment.appUriScheme) {
                    return extHostUrls.createAppUri(uri);
                }
                try {
                    return await extHostWindow.asExternalUri(uri, { allowTunneling: !!initData.remote.authority });
                }
                catch (err) {
                    if (matchesScheme(uri, Schemas.http) || matchesScheme(uri, Schemas.https)) {
                        return uri;
                    }
                    throw err;
                }
            },
            get remoteName() {
                return getRemoteName(initData.remote.authority);
            },
            get remoteAuthority() {
                checkProposedApiEnabled(extension, 'resolvers');
                return initData.remote.authority;
            },
            get uiKind() {
                return initData.uiKind;
            },
            get logLevel() {
                return extHostLogService.getLevel();
            },
            get onDidChangeLogLevel() {
                return extHostLogService.onDidChangeLogLevel;
            },
            registerIssueUriRequestHandler(handler) {
                checkProposedApiEnabled(extension, 'handleIssueUri');
                return extHostIssueReporter.registerIssueUriRequestHandler(extension, handler);
            },
            get appQuality() {
                checkProposedApiEnabled(extension, 'resolvers');
                return initData.quality;
            },
            get appCommit() {
                checkProposedApiEnabled(extension, 'resolvers');
                return initData.commit;
            },
        };
        const tests = {
            createTestController(provider, label, refreshHandler) {
                return extHostTesting.createTestController(extension, provider, label, refreshHandler);
            },
            createTestObserver() {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.createTestObserver();
            },
            runTests(provider) {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.runTests(provider);
            },
            get onDidChangeTestResults() {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.onResultsChanged;
            },
            get testResults() {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.results;
            },
        };
        const extensionKind = initData.remote.isRemote
            ? ExtensionKind.Workspace
            : ExtensionKind.UI;
        const extensions = {
            getExtension(extensionId, includeFromDifferentExtensionHosts) {
                if (!isProposedApiEnabled(extension, 'extensionsAny')) {
                    includeFromDifferentExtensionHosts = false;
                }
                const mine = extensionInfo.mine.getExtensionDescription(extensionId);
                if (mine) {
                    return ( new Extension(extensionService, extension.identifier, mine, extensionKind, false));
                }
                if (includeFromDifferentExtensionHosts) {
                    const foreign = extensionInfo.all.getExtensionDescription(extensionId);
                    if (foreign) {
                        return ( new Extension(
                            extensionService,
                            extension.identifier,
                            foreign,
                            extensionKind ,
                            true
                        ));
                    }
                }
                return undefined;
            },
            get all() {
                const result = [];
                for (const desc of extensionInfo.mine.getAllExtensionDescriptions()) {
                    result.push(( new Extension(extensionService, extension.identifier, desc, extensionKind, false)));
                }
                return result;
            },
            get allAcrossExtensionHosts() {
                checkProposedApiEnabled(extension, 'extensionsAny');
                const local = ( new ExtensionIdentifierSet(( extensionInfo.mine.getAllExtensionDescriptions().map(desc => desc.identifier))));
                const result = [];
                for (const desc of extensionInfo.all.getAllExtensionDescriptions()) {
                    const isFromDifferentExtensionHost = !( local.has(desc.identifier));
                    result.push(( new Extension(
                        extensionService,
                        extension.identifier,
                        desc,
                        extensionKind ,
                        isFromDifferentExtensionHost
                    )));
                }
                return result;
            },
            get onDidChange() {
                if (isProposedApiEnabled(extension, 'extensionsAny')) {
                    return Event.any(extensionInfo.mine.onDidChange, extensionInfo.all.onDidChange);
                }
                return extensionInfo.mine.onDidChange;
            }
        };
        const languages = {
            createDiagnosticCollection(name) {
                return extHostDiagnostics.createDiagnosticCollection(extension.identifier, name);
            },
            get onDidChangeDiagnostics() {
                return extHostDiagnostics.onDidChangeDiagnostics;
            },
            getDiagnostics: (resource) => {
                return extHostDiagnostics.getDiagnostics(resource);
            },
            getLanguages() {
                return extHostLanguages.getLanguages();
            },
            setTextDocumentLanguage(document, languageId) {
                return extHostLanguages.changeLanguage(document.uri, languageId);
            },
            match(selector, document) {
                const notebook = extHostDocuments.getDocumentData(document.uri)?.notebook;
                return score(LanguageSelector.from(selector), document.uri, document.languageId, true, notebook?.uri, notebook?.notebookType);
            },
            registerCodeActionsProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerCodeActionProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerDocumentPasteEditProvider(selector, provider, metadata) {
                checkProposedApiEnabled(extension, 'documentPaste');
                return extHostLanguageFeatures.registerDocumentPasteEditProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerCodeLensProvider(selector, provider) {
                return extHostLanguageFeatures.registerCodeLensProvider(extension, checkSelector(selector), provider);
            },
            registerDefinitionProvider(selector, provider) {
                return extHostLanguageFeatures.registerDefinitionProvider(extension, checkSelector(selector), provider);
            },
            registerDeclarationProvider(selector, provider) {
                return extHostLanguageFeatures.registerDeclarationProvider(extension, checkSelector(selector), provider);
            },
            registerImplementationProvider(selector, provider) {
                return extHostLanguageFeatures.registerImplementationProvider(extension, checkSelector(selector), provider);
            },
            registerTypeDefinitionProvider(selector, provider) {
                return extHostLanguageFeatures.registerTypeDefinitionProvider(extension, checkSelector(selector), provider);
            },
            registerHoverProvider(selector, provider) {
                return extHostLanguageFeatures.registerHoverProvider(extension, checkSelector(selector), provider, extension.identifier);
            },
            registerEvaluatableExpressionProvider(selector, provider) {
                return extHostLanguageFeatures.registerEvaluatableExpressionProvider(extension, checkSelector(selector), provider, extension.identifier);
            },
            registerInlineValuesProvider(selector, provider) {
                return extHostLanguageFeatures.registerInlineValuesProvider(extension, checkSelector(selector), provider, extension.identifier);
            },
            registerDocumentHighlightProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentHighlightProvider(extension, checkSelector(selector), provider);
            },
            registerLinkedEditingRangeProvider(selector, provider) {
                return extHostLanguageFeatures.registerLinkedEditingRangeProvider(extension, checkSelector(selector), provider);
            },
            registerReferenceProvider(selector, provider) {
                return extHostLanguageFeatures.registerReferenceProvider(extension, checkSelector(selector), provider);
            },
            registerRenameProvider(selector, provider) {
                return extHostLanguageFeatures.registerRenameProvider(extension, checkSelector(selector), provider);
            },
            registerDocumentSymbolProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerDocumentSymbolProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerWorkspaceSymbolProvider(provider) {
                return extHostLanguageFeatures.registerWorkspaceSymbolProvider(extension, provider);
            },
            registerDocumentFormattingEditProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentFormattingEditProvider(extension, checkSelector(selector), provider);
            },
            registerDocumentRangeFormattingEditProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentRangeFormattingEditProvider(extension, checkSelector(selector), provider);
            },
            registerOnTypeFormattingEditProvider(selector, provider, firstTriggerCharacter, ...moreTriggerCharacters) {
                return extHostLanguageFeatures.registerOnTypeFormattingEditProvider(extension, checkSelector(selector), provider, [firstTriggerCharacter].concat(moreTriggerCharacters));
            },
            registerDocumentSemanticTokensProvider(selector, provider, legend) {
                return extHostLanguageFeatures.registerDocumentSemanticTokensProvider(extension, checkSelector(selector), provider, legend);
            },
            registerDocumentRangeSemanticTokensProvider(selector, provider, legend) {
                return extHostLanguageFeatures.registerDocumentRangeSemanticTokensProvider(extension, checkSelector(selector), provider, legend);
            },
            registerSignatureHelpProvider(selector, provider, firstItem, ...remaining) {
                if (typeof firstItem === 'object') {
                    return extHostLanguageFeatures.registerSignatureHelpProvider(extension, checkSelector(selector), provider, firstItem);
                }
                return extHostLanguageFeatures.registerSignatureHelpProvider(extension, checkSelector(selector), provider, typeof firstItem === 'undefined' ? [] : [firstItem, ...remaining]);
            },
            registerCompletionItemProvider(selector, provider, ...triggerCharacters) {
                return extHostLanguageFeatures.registerCompletionItemProvider(extension, checkSelector(selector), provider, triggerCharacters);
            },
            registerInlineCompletionItemProvider(selector, provider, metadata) {
                if (provider.handleDidShowCompletionItem) {
                    checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                }
                if (provider.handleDidPartiallyAcceptCompletionItem) {
                    checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                }
                if (metadata) {
                    checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                }
                return extHostLanguageFeatures.registerInlineCompletionsProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerDocumentLinkProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentLinkProvider(extension, checkSelector(selector), provider);
            },
            registerColorProvider(selector, provider) {
                return extHostLanguageFeatures.registerColorProvider(extension, checkSelector(selector), provider);
            },
            registerFoldingRangeProvider(selector, provider) {
                return extHostLanguageFeatures.registerFoldingRangeProvider(extension, checkSelector(selector), provider);
            },
            registerSelectionRangeProvider(selector, provider) {
                return extHostLanguageFeatures.registerSelectionRangeProvider(extension, selector, provider);
            },
            registerCallHierarchyProvider(selector, provider) {
                return extHostLanguageFeatures.registerCallHierarchyProvider(extension, selector, provider);
            },
            registerTypeHierarchyProvider(selector, provider) {
                return extHostLanguageFeatures.registerTypeHierarchyProvider(extension, selector, provider);
            },
            setLanguageConfiguration: (language, configuration) => {
                return extHostLanguageFeatures.setLanguageConfiguration(extension, language, configuration);
            },
            getTokenInformationAtPosition(doc, pos) {
                checkProposedApiEnabled(extension, 'tokenInformation');
                return extHostLanguages.tokenAtPosition(doc, pos);
            },
            registerInlayHintsProvider(selector, provider) {
                return extHostLanguageFeatures.registerInlayHintsProvider(extension, selector, provider);
            },
            createLanguageStatusItem(id, selector) {
                return extHostLanguages.createLanguageStatusItem(extension, id, selector);
            },
            registerDocumentDropEditProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerDocumentOnDropEditProvider(extension, selector, provider, isProposedApiEnabled(extension, 'dropMetadata') ? metadata : undefined);
            }
        };
        const window = {
            get activeTextEditor() {
                return extHostEditors.getActiveTextEditor();
            },
            get visibleTextEditors() {
                return extHostEditors.getVisibleTextEditors();
            },
            get activeTerminal() {
                return extHostTerminalService.activeTerminal;
            },
            get terminals() {
                return extHostTerminalService.terminals;
            },
            async showTextDocument(documentOrUri, columnOrOptions, preserveFocus) {
                const document = await (URI.isUri(documentOrUri)
                    ? Promise.resolve(workspace.openTextDocument(documentOrUri))
                    : Promise.resolve(documentOrUri));
                return extHostEditors.showTextDocument(document, columnOrOptions, preserveFocus);
            },
            createTextEditorDecorationType(options) {
                return extHostEditors.createTextEditorDecorationType(extension, options);
            },
            onDidChangeActiveTextEditor(listener, thisArg, disposables) {
                return extHostEditors.onDidChangeActiveTextEditor(listener, thisArg, disposables);
            },
            onDidChangeVisibleTextEditors(listener, thisArg, disposables) {
                return extHostEditors.onDidChangeVisibleTextEditors(listener, thisArg, disposables);
            },
            onDidChangeTextEditorSelection(listener, thisArgs, disposables) {
                return extHostEditors.onDidChangeTextEditorSelection(listener, thisArgs, disposables);
            },
            onDidChangeTextEditorOptions(listener, thisArgs, disposables) {
                return extHostEditors.onDidChangeTextEditorOptions(listener, thisArgs, disposables);
            },
            onDidChangeTextEditorVisibleRanges(listener, thisArgs, disposables) {
                return extHostEditors.onDidChangeTextEditorVisibleRanges(listener, thisArgs, disposables);
            },
            onDidChangeTextEditorViewColumn(listener, thisArg, disposables) {
                return extHostEditors.onDidChangeTextEditorViewColumn(listener, thisArg, disposables);
            },
            onDidCloseTerminal(listener, thisArg, disposables) {
                return extHostTerminalService.onDidCloseTerminal(listener, thisArg, disposables);
            },
            onDidOpenTerminal(listener, thisArg, disposables) {
                return extHostTerminalService.onDidOpenTerminal(listener, thisArg, disposables);
            },
            onDidChangeActiveTerminal(listener, thisArg, disposables) {
                return extHostTerminalService.onDidChangeActiveTerminal(listener, thisArg, disposables);
            },
            onDidChangeTerminalDimensions(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'terminalDimensions');
                return extHostTerminalService.onDidChangeTerminalDimensions(listener, thisArg, disposables);
            },
            onDidChangeTerminalState(listener, thisArg, disposables) {
                return extHostTerminalService.onDidChangeTerminalState(listener, thisArg, disposables);
            },
            onDidWriteTerminalData(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'terminalDataWriteEvent');
                return extHostTerminalService.onDidWriteTerminalData(listener, thisArg, disposables);
            },
            onDidExecuteTerminalCommand(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'terminalExecuteCommandEvent');
                return extHostTerminalService.onDidExecuteTerminalCommand(listener, thisArg, disposables);
            },
            get state() {
                return extHostWindow.getState(extension);
            },
            onDidChangeWindowState(listener, thisArg, disposables) {
                return extHostWindow.onDidChangeWindowState(listener, thisArg, disposables);
            },
            showInformationMessage(message, ...rest) {
                return extHostMessageService.showMessage(extension, Severity.Info, message, rest[0], rest.slice(1));
            },
            showWarningMessage(message, ...rest) {
                return extHostMessageService.showMessage(extension, Severity.Warning, message, rest[0], rest.slice(1));
            },
            showErrorMessage(message, ...rest) {
                return extHostMessageService.showMessage(extension, Severity.Error, message, rest[0], rest.slice(1));
            },
            showQuickPick(items, options, token) {
                return extHostQuickOpen.showQuickPick(extension, items, options, token);
            },
            showWorkspaceFolderPick(options) {
                return extHostQuickOpen.showWorkspaceFolderPick(options);
            },
            showInputBox(options, token) {
                return extHostQuickOpen.showInput(options, token);
            },
            showOpenDialog(options) {
                return extHostDialogs.showOpenDialog(extension, options);
            },
            showSaveDialog(options) {
                return extHostDialogs.showSaveDialog(options);
            },
            createStatusBarItem(alignmentOrId, priorityOrAlignment, priorityArg) {
                let id;
                let alignment;
                let priority;
                if (typeof alignmentOrId === 'string') {
                    id = alignmentOrId;
                    alignment = priorityOrAlignment;
                    priority = priorityArg;
                }
                else {
                    alignment = alignmentOrId;
                    priority = priorityOrAlignment;
                }
                return extHostStatusBar.createStatusBarEntry(extension, id, alignment, priority);
            },
            setStatusBarMessage(text, timeoutOrThenable) {
                return extHostStatusBar.setStatusBarMessage(text, timeoutOrThenable);
            },
            withScmProgress(task) {
                extHostApiDeprecation.report('window.withScmProgress', extension, `Use 'withProgress' instead.`);
                return extHostProgress.withProgress(extension, { location: ProgressLocation.SourceControl }, (progress, token) => task({ report(n) { } }));
            },
            withProgress(options, task) {
                return extHostProgress.withProgress(extension, options, task);
            },
            createOutputChannel(name, options) {
                return extHostOutputService.createOutputChannel(name, options, extension);
            },
            createWebviewPanel(viewType, title, showOptions, options) {
                return extHostWebviewPanels.createWebviewPanel(extension, viewType, title, showOptions, options);
            },
            createWebviewTextEditorInset(editor, line, height, options) {
                checkProposedApiEnabled(extension, 'editorInsets');
                return extHostEditorInsets.createWebviewEditorInset(editor, line, height, options, extension);
            },
            createTerminal(nameOrOptions, shellPath, shellArgs) {
                if (typeof nameOrOptions === 'object') {
                    if ('pty' in nameOrOptions) {
                        return extHostTerminalService.createExtensionTerminal(nameOrOptions);
                    }
                    return extHostTerminalService.createTerminalFromOptions(nameOrOptions);
                }
                return extHostTerminalService.createTerminal(nameOrOptions, shellPath, shellArgs);
            },
            registerTerminalLinkProvider(provider) {
                return extHostTerminalService.registerLinkProvider(provider);
            },
            registerTerminalProfileProvider(id, provider) {
                return extHostTerminalService.registerProfileProvider(extension, id, provider);
            },
            registerTerminalQuickFixProvider(id, provider) {
                checkProposedApiEnabled(extension, 'terminalQuickFixProvider');
                return extHostTerminalService.registerTerminalQuickFixProvider(id, extension.identifier.value, provider);
            },
            registerTreeDataProvider(viewId, treeDataProvider) {
                return extHostTreeViews.registerTreeDataProvider(viewId, treeDataProvider, extension);
            },
            createTreeView(viewId, options) {
                return extHostTreeViews.createTreeView(viewId, options, extension);
            },
            registerWebviewPanelSerializer: (viewType, serializer) => {
                return extHostWebviewPanels.registerWebviewPanelSerializer(extension, viewType, serializer);
            },
            registerCustomEditorProvider: (viewType, provider, options = {}) => {
                return extHostCustomEditors.registerCustomEditorProvider(extension, viewType, provider, options);
            },
            registerFileDecorationProvider(provider) {
                return extHostDecorations.registerFileDecorationProvider(provider, extension);
            },
            registerUriHandler(handler) {
                return extHostUrls.registerUriHandler(extension, handler);
            },
            createQuickPick() {
                return extHostQuickOpen.createQuickPick(extension);
            },
            createInputBox() {
                return extHostQuickOpen.createInputBox(extension);
            },
            get activeColorTheme() {
                return extHostTheming.activeColorTheme;
            },
            onDidChangeActiveColorTheme(listener, thisArg, disposables) {
                return extHostTheming.onDidChangeActiveColorTheme(listener, thisArg, disposables);
            },
            registerWebviewViewProvider(viewId, provider, options) {
                return extHostWebviewViews.registerWebviewViewProvider(extension, viewId, provider, options?.webviewOptions);
            },
            get activeNotebookEditor() {
                return extHostNotebook.activeNotebookEditor;
            },
            onDidChangeActiveNotebookEditor(listener, thisArgs, disposables) {
                return extHostNotebook.onDidChangeActiveNotebookEditor(listener, thisArgs, disposables);
            },
            get visibleNotebookEditors() {
                return extHostNotebook.visibleNotebookEditors;
            },
            get onDidChangeVisibleNotebookEditors() {
                return extHostNotebook.onDidChangeVisibleNotebookEditors;
            },
            onDidChangeNotebookEditorSelection(listener, thisArgs, disposables) {
                return extHostNotebookEditors.onDidChangeNotebookEditorSelection(listener, thisArgs, disposables);
            },
            onDidChangeNotebookEditorVisibleRanges(listener, thisArgs, disposables) {
                return extHostNotebookEditors.onDidChangeNotebookEditorVisibleRanges(listener, thisArgs, disposables);
            },
            showNotebookDocument(document, options) {
                return extHostNotebook.showNotebookDocument(document, options);
            },
            registerExternalUriOpener(id, opener, metadata) {
                checkProposedApiEnabled(extension, 'externalUriOpener');
                return extHostUriOpeners.registerExternalUriOpener(extension.identifier, id, opener, metadata);
            },
            registerProfileContentHandler(id, handler) {
                checkProposedApiEnabled(extension, 'profileContentHandlers');
                return extHostProfileContentHandlers.registrProfileContentHandler(extension, id, handler);
            },
            registerQuickDiffProvider(selector, quickDiffProvider, label, rootUri) {
                checkProposedApiEnabled(extension, 'quickDiffProvider');
                return extHostQuickDiff.registerQuickDiffProvider(checkSelector(selector), quickDiffProvider, label, rootUri);
            },
            get tabGroups() {
                return extHostEditorTabs.tabGroups;
            },
            registerShareProvider(selector, provider) {
                checkProposedApiEnabled(extension, 'shareProvider');
                return extHostShare.registerShareProvider(checkSelector(selector), provider);
            }
        };
        const workspace = {
            get rootPath() {
                extHostApiDeprecation.report('workspace.rootPath', extension, `Please use 'workspace.workspaceFolders' instead. More details: https://aka.ms/vscode-eliminating-rootpath`);
                return extHostWorkspace.getPath();
            },
            set rootPath(value) {
                throw new errors.ReadonlyError('rootPath');
            },
            getWorkspaceFolder(resource) {
                return extHostWorkspace.getWorkspaceFolder(resource);
            },
            get workspaceFolders() {
                return extHostWorkspace.getWorkspaceFolders();
            },
            get name() {
                return extHostWorkspace.name;
            },
            set name(value) {
                throw new errors.ReadonlyError('name');
            },
            get workspaceFile() {
                return extHostWorkspace.workspaceFile;
            },
            set workspaceFile(value) {
                throw new errors.ReadonlyError('workspaceFile');
            },
            updateWorkspaceFolders: (index, deleteCount, ...workspaceFoldersToAdd) => {
                return extHostWorkspace.updateWorkspaceFolders(extension, index, deleteCount || 0, ...workspaceFoldersToAdd);
            },
            onDidChangeWorkspaceFolders: function (listener, thisArgs, disposables) {
                return extHostWorkspace.onDidChangeWorkspace(listener, thisArgs, disposables);
            },
            asRelativePath: (pathOrUri, includeWorkspace) => {
                return extHostWorkspace.getRelativePath(pathOrUri, includeWorkspace);
            },
            findFiles: (include, exclude, maxResults, token) => {
                return extHostWorkspace.findFiles(include, exclude, maxResults, extension.identifier, token);
            },
            findTextInFiles: (query, optionsOrCallback, callbackOrToken, token) => {
                checkProposedApiEnabled(extension, 'findTextInFiles');
                let options;
                let callback;
                if (typeof optionsOrCallback === 'object') {
                    options = optionsOrCallback;
                    callback = callbackOrToken;
                }
                else {
                    options = {};
                    callback = optionsOrCallback;
                    token = callbackOrToken;
                }
                return extHostWorkspace.findTextInFiles(query, options || {}, callback, extension.identifier, token);
            },
            save: (uri) => {
                checkProposedApiEnabled(extension, 'saveEditor');
                return extHostWorkspace.save(uri);
            },
            saveAs: (uri) => {
                checkProposedApiEnabled(extension, 'saveEditor');
                return extHostWorkspace.saveAs(uri);
            },
            saveAll: (includeUntitled) => {
                return extHostWorkspace.saveAll(includeUntitled);
            },
            applyEdit(edit, metadata) {
                return extHostBulkEdits.applyWorkspaceEdit(edit, extension, metadata);
            },
            createFileSystemWatcher: (pattern, ignoreCreate, ignoreChange, ignoreDelete) => {
                return extHostFileSystemEvent.createFileSystemWatcher(extHostWorkspace, extension, pattern, ignoreCreate, ignoreChange, ignoreDelete);
            },
            get textDocuments() {
                return ( extHostDocuments.getAllDocumentData().map(data => data.document));
            },
            set textDocuments(value) {
                throw new errors.ReadonlyError('textDocuments');
            },
            openTextDocument(uriOrFileNameOrOptions) {
                let uriPromise;
                const options = uriOrFileNameOrOptions;
                if (typeof uriOrFileNameOrOptions === 'string') {
                    uriPromise = Promise.resolve(URI.file(uriOrFileNameOrOptions));
                }
                else if (URI.isUri(uriOrFileNameOrOptions)) {
                    uriPromise = Promise.resolve(uriOrFileNameOrOptions);
                }
                else if (!options || typeof options === 'object') {
                    uriPromise = extHostDocuments.createDocumentData(options);
                }
                else {
                    throw new Error('illegal argument - uriOrFileNameOrOptions');
                }
                return uriPromise.then(uri => {
                    return extHostDocuments.ensureDocumentData(uri).then(documentData => {
                        return documentData.document;
                    });
                });
            },
            onDidOpenTextDocument: (listener, thisArgs, disposables) => {
                return extHostDocuments.onDidAddDocument(listener, thisArgs, disposables);
            },
            onDidCloseTextDocument: (listener, thisArgs, disposables) => {
                return extHostDocuments.onDidRemoveDocument(listener, thisArgs, disposables);
            },
            onDidChangeTextDocument: (listener, thisArgs, disposables) => {
                return extHostDocuments.onDidChangeDocument(listener, thisArgs, disposables);
            },
            onDidSaveTextDocument: (listener, thisArgs, disposables) => {
                return extHostDocuments.onDidSaveDocument(listener, thisArgs, disposables);
            },
            onWillSaveTextDocument: (listener, thisArgs, disposables) => {
                return extHostDocumentSaveParticipant.getOnWillSaveTextDocumentEvent(extension)(listener, thisArgs, disposables);
            },
            get notebookDocuments() {
                return ( extHostNotebook.notebookDocuments.map(d => d.apiNotebook));
            },
            async openNotebookDocument(uriOrType, content) {
                let uri;
                if (URI.isUri(uriOrType)) {
                    uri = uriOrType;
                    await extHostNotebook.openNotebookDocument(uriOrType);
                }
                else if (typeof uriOrType === 'string') {
                    uri = URI.revive(await extHostNotebook.createNotebookDocument({ viewType: uriOrType, content }));
                }
                else {
                    throw new Error('Invalid arguments');
                }
                return extHostNotebook.getNotebookDocument(uri).apiNotebook;
            },
            onDidSaveNotebookDocument(listener, thisArg, disposables) {
                return extHostNotebookDocuments.onDidSaveNotebookDocument(listener, thisArg, disposables);
            },
            onDidChangeNotebookDocument(listener, thisArg, disposables) {
                return extHostNotebookDocuments.onDidChangeNotebookDocument(listener, thisArg, disposables);
            },
            onWillSaveNotebookDocument(listener, thisArg, disposables) {
                return extHostNotebookDocumentSaveParticipant.getOnWillSaveNotebookDocumentEvent(extension)(listener, thisArg, disposables);
            },
            get onDidOpenNotebookDocument() {
                return extHostNotebook.onDidOpenNotebookDocument;
            },
            get onDidCloseNotebookDocument() {
                return extHostNotebook.onDidCloseNotebookDocument;
            },
            registerNotebookSerializer(viewType, serializer, options, registration) {
                return extHostNotebook.registerNotebookSerializer(extension, viewType, serializer, options, isProposedApiEnabled(extension, 'notebookLiveShare') ? registration : undefined);
            },
            onDidChangeConfiguration: (listener, thisArgs, disposables) => {
                return configProvider.onDidChangeConfiguration(listener, thisArgs, disposables);
            },
            getConfiguration(section, scope) {
                scope = arguments.length === 1 ? undefined : scope;
                return configProvider.getConfiguration(section, scope, extension);
            },
            registerTextDocumentContentProvider(scheme, provider) {
                return extHostDocumentContentProviders.registerTextDocumentContentProvider(scheme, provider);
            },
            registerTaskProvider: (type, provider) => {
                extHostApiDeprecation.report('window.registerTaskProvider', extension, `Use the corresponding function on the 'tasks' namespace instead`);
                return extHostTask.registerTaskProvider(extension, type, provider);
            },
            registerFileSystemProvider(scheme, provider, options) {
                return combinedDisposable(extHostFileSystem.registerFileSystemProvider(extension, scheme, provider, options), extHostConsumerFileSystem.addFileSystemProvider(scheme, provider, options));
            },
            get fs() {
                return extHostConsumerFileSystem.value;
            },
            registerFileSearchProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'fileSearchProvider');
                return extHostSearch.registerFileSearchProvider(scheme, provider);
            },
            registerTextSearchProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'textSearchProvider');
                return extHostSearch.registerTextSearchProvider(scheme, provider);
            },
            registerRemoteAuthorityResolver: (authorityPrefix, resolver) => {
                checkProposedApiEnabled(extension, 'resolvers');
                return extensionService.registerRemoteAuthorityResolver(authorityPrefix, resolver);
            },
            registerResourceLabelFormatter: (formatter) => {
                checkProposedApiEnabled(extension, 'resolvers');
                return extHostLabelService.$registerResourceLabelFormatter(formatter);
            },
            getRemoteExecServer: (authority) => {
                checkProposedApiEnabled(extension, 'resolvers');
                return extensionService.getRemoteExecServer(authority);
            },
            onDidCreateFiles: (listener, thisArg, disposables) => {
                return extHostFileSystemEvent.onDidCreateFile(listener, thisArg, disposables);
            },
            onDidDeleteFiles: (listener, thisArg, disposables) => {
                return extHostFileSystemEvent.onDidDeleteFile(listener, thisArg, disposables);
            },
            onDidRenameFiles: (listener, thisArg, disposables) => {
                return extHostFileSystemEvent.onDidRenameFile(listener, thisArg, disposables);
            },
            onWillCreateFiles: (listener, thisArg, disposables) => {
                return extHostFileSystemEvent.getOnWillCreateFileEvent(extension)(listener, thisArg, disposables);
            },
            onWillDeleteFiles: (listener, thisArg, disposables) => {
                return extHostFileSystemEvent.getOnWillDeleteFileEvent(extension)(listener, thisArg, disposables);
            },
            onWillRenameFiles: (listener, thisArg, disposables) => {
                return extHostFileSystemEvent.getOnWillRenameFileEvent(extension)(listener, thisArg, disposables);
            },
            openTunnel: (forward) => {
                checkProposedApiEnabled(extension, 'tunnels');
                return extHostTunnelService.openTunnel(extension, forward).then(value => {
                    if (!value) {
                        throw new Error('cannot open tunnel');
                    }
                    return value;
                });
            },
            get tunnels() {
                checkProposedApiEnabled(extension, 'tunnels');
                return extHostTunnelService.getTunnels();
            },
            onDidChangeTunnels: (listener, thisArg, disposables) => {
                checkProposedApiEnabled(extension, 'tunnels');
                return extHostTunnelService.onDidChangeTunnels(listener, thisArg, disposables);
            },
            registerPortAttributesProvider: (portSelector, provider) => {
                checkProposedApiEnabled(extension, 'portsAttributes');
                return extHostTunnelService.registerPortsAttributesProvider(portSelector, provider);
            },
            registerTunnelProvider: (tunnelProvider, information) => {
                checkProposedApiEnabled(extension, 'tunnelFactory');
                return extHostTunnelService.registerTunnelProvider(tunnelProvider, information);
            },
            registerTimelineProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'timeline');
                return extHostTimeline.registerTimelineProvider(scheme, provider, extension.identifier, extHostCommands.converter);
            },
            get isTrusted() {
                return extHostWorkspace.trusted;
            },
            requestWorkspaceTrust: (options) => {
                checkProposedApiEnabled(extension, 'workspaceTrust');
                return extHostWorkspace.requestWorkspaceTrust(options);
            },
            onDidGrantWorkspaceTrust: (listener, thisArgs, disposables) => {
                return extHostWorkspace.onDidGrantWorkspaceTrust(listener, thisArgs, disposables);
            },
            registerEditSessionIdentityProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'editSessionIdentityProvider');
                return extHostWorkspace.registerEditSessionIdentityProvider(scheme, provider);
            },
            onWillCreateEditSessionIdentity: (listener, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'editSessionIdentityProvider');
                return extHostWorkspace.getOnWillCreateEditSessionIdentityEvent(extension)(listener, thisArgs, disposables);
            },
            registerCanonicalUriProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'canonicalUriProvider');
                return extHostWorkspace.registerCanonicalUriProvider(scheme, provider);
            },
            getCanonicalUri: (uri, options, token) => {
                checkProposedApiEnabled(extension, 'canonicalUriProvider');
                return extHostWorkspace.provideCanonicalUri(uri, options, token);
            }
        };
        const scm = {
            get inputBox() {
                extHostApiDeprecation.report('scm.inputBox', extension, `Use 'SourceControl.inputBox' instead`);
                return extHostSCM.getLastInputBox(extension);
            },
            createSourceControl(id, label, rootUri) {
                return extHostSCM.createSourceControl(extension, id, label, rootUri);
            }
        };
        const comments = {
            createCommentController(id, label) {
                return extHostComment.createCommentController(extension, id, label);
            }
        };
        const debug = {
            get activeDebugSession() {
                return extHostDebugService.activeDebugSession;
            },
            get activeDebugConsole() {
                return extHostDebugService.activeDebugConsole;
            },
            get breakpoints() {
                return extHostDebugService.breakpoints;
            },
            get stackFrameFocus() {
                return extHostDebugService.stackFrameFocus;
            },
            onDidStartDebugSession(listener, thisArg, disposables) {
                return extHostDebugService.onDidStartDebugSession(listener, thisArg, disposables);
            },
            onDidTerminateDebugSession(listener, thisArg, disposables) {
                return extHostDebugService.onDidTerminateDebugSession(listener, thisArg, disposables);
            },
            onDidChangeActiveDebugSession(listener, thisArg, disposables) {
                return extHostDebugService.onDidChangeActiveDebugSession(listener, thisArg, disposables);
            },
            onDidReceiveDebugSessionCustomEvent(listener, thisArg, disposables) {
                return extHostDebugService.onDidReceiveDebugSessionCustomEvent(listener, thisArg, disposables);
            },
            onDidChangeBreakpoints(listener, thisArgs, disposables) {
                return extHostDebugService.onDidChangeBreakpoints(listener, thisArgs, disposables);
            },
            onDidChangeStackFrameFocus(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'debugFocus');
                return extHostDebugService.onDidChangeStackFrameFocus(listener, thisArg, disposables);
            },
            registerDebugConfigurationProvider(debugType, provider, triggerKind) {
                return extHostDebugService.registerDebugConfigurationProvider(debugType, provider, triggerKind || DebugConfigurationProviderTriggerKind.Initial);
            },
            registerDebugAdapterDescriptorFactory(debugType, factory) {
                return extHostDebugService.registerDebugAdapterDescriptorFactory(extension, debugType, factory);
            },
            registerDebugAdapterTrackerFactory(debugType, factory) {
                return extHostDebugService.registerDebugAdapterTrackerFactory(debugType, factory);
            },
            startDebugging(folder, nameOrConfig, parentSessionOrOptions) {
                if (!parentSessionOrOptions || (typeof parentSessionOrOptions === 'object' && 'configuration' in parentSessionOrOptions)) {
                    return extHostDebugService.startDebugging(folder, nameOrConfig, { parentSession: parentSessionOrOptions });
                }
                return extHostDebugService.startDebugging(folder, nameOrConfig, parentSessionOrOptions || {});
            },
            stopDebugging(session) {
                return extHostDebugService.stopDebugging(session);
            },
            addBreakpoints(breakpoints) {
                return extHostDebugService.addBreakpoints(breakpoints);
            },
            removeBreakpoints(breakpoints) {
                return extHostDebugService.removeBreakpoints(breakpoints);
            },
            asDebugSourceUri(source, session) {
                return extHostDebugService.asDebugSourceUri(source, session);
            }
        };
        const tasks = {
            registerTaskProvider: (type, provider) => {
                return extHostTask.registerTaskProvider(extension, type, provider);
            },
            fetchTasks: (filter) => {
                return extHostTask.fetchTasks(filter);
            },
            executeTask: (task) => {
                return extHostTask.executeTask(extension, task);
            },
            get taskExecutions() {
                return extHostTask.taskExecutions;
            },
            onDidStartTask: (listeners, thisArgs, disposables) => {
                return extHostTask.onDidStartTask(listeners, thisArgs, disposables);
            },
            onDidEndTask: (listeners, thisArgs, disposables) => {
                return extHostTask.onDidEndTask(listeners, thisArgs, disposables);
            },
            onDidStartTaskProcess: (listeners, thisArgs, disposables) => {
                return extHostTask.onDidStartTaskProcess(listeners, thisArgs, disposables);
            },
            onDidEndTaskProcess: (listeners, thisArgs, disposables) => {
                return extHostTask.onDidEndTaskProcess(listeners, thisArgs, disposables);
            }
        };
        const notebooks = {
            createNotebookController(id, notebookType, label, handler, rendererScripts) {
                return extHostNotebookKernels.createNotebookController(extension, id, notebookType, label, handler, isProposedApiEnabled(extension, 'notebookMessaging') ? rendererScripts : undefined);
            },
            registerNotebookCellStatusBarItemProvider: (notebookType, provider) => {
                return extHostNotebook.registerNotebookCellStatusBarItemProvider(extension, notebookType, provider);
            },
            createRendererMessaging(rendererId) {
                return extHostNotebookRenderers.createRendererMessaging(extension, rendererId);
            },
            createNotebookControllerDetectionTask(notebookType) {
                checkProposedApiEnabled(extension, 'notebookKernelSource');
                return extHostNotebookKernels.createNotebookControllerDetectionTask(extension, notebookType);
            },
            registerKernelSourceActionProvider(notebookType, provider) {
                checkProposedApiEnabled(extension, 'notebookKernelSource');
                return extHostNotebookKernels.registerKernelSourceActionProvider(extension, notebookType, provider);
            },
            onDidChangeNotebookCellExecutionState(listener, thisArgs, disposables) {
                checkProposedApiEnabled(extension, 'notebookCellExecutionState');
                return extHostNotebookKernels.onDidChangeNotebookCellExecutionState(listener, thisArgs, disposables);
            }
        };
        const l10n = {
            t(...params) {
                if (typeof params[0] === 'string') {
                    const key = params.shift();
                    const argsFormatted = !params || typeof params[0] !== 'object' ? params : params[0];
                    return extHostLocalization.getMessage(extension.identifier.value, { message: key, args: argsFormatted });
                }
                return extHostLocalization.getMessage(extension.identifier.value, params[0]);
            },
            get bundle() {
                return extHostLocalization.getBundle(extension.identifier.value);
            },
            get uri() {
                return extHostLocalization.getBundleUri(extension.identifier.value);
            }
        };
        const interactive = {
            _version: 1,
            registerInteractiveEditorSessionProvider(provider, metadata) {
                checkProposedApiEnabled(extension, 'interactive');
                return extHostInteractiveEditor.registerProvider(extension, provider, metadata = { label: metadata?.label ?? extension.displayName ?? extension.name });
            },
            registerInteractiveSessionProvider(id, provider) {
                checkProposedApiEnabled(extension, 'interactive');
                return extHostChat.registerChatProvider(extension, id, provider);
            },
            addInteractiveRequest(context) {
                checkProposedApiEnabled(extension, 'interactive');
                return extHostChat.addChatRequest(context);
            },
            sendInteractiveRequestToProvider(providerId, message) {
                checkProposedApiEnabled(extension, 'interactive');
                return extHostChat.sendInteractiveRequestToProvider(providerId, message);
            },
            get onDidPerformUserAction() {
                checkProposedApiEnabled(extension, 'interactiveUserActions');
                return extHostChat.onDidPerformUserAction;
            },
            transferChatSession(session, toWorkspace) {
                checkProposedApiEnabled(extension, 'interactive');
                return extHostChat.transferChatSession(session, toWorkspace);
            }
        };
        const ai = {
            getRelatedInformation(query, types) {
                checkProposedApiEnabled(extension, 'aiRelatedInformation');
                return extHostAiRelatedInformation.getRelatedInformation(extension, query, types);
            },
            registerRelatedInformationProvider(type, provider) {
                checkProposedApiEnabled(extension, 'aiRelatedInformation');
                return extHostAiRelatedInformation.registerRelatedInformationProvider(extension, type, provider);
            },
            registerEmbeddingVectorProvider(model, provider) {
                checkProposedApiEnabled(extension, 'aiRelatedInformation');
                return extHostAiEmbeddingVector.registerEmbeddingVectorProvider(extension, model, provider);
            }
        };
        const chat = {
            registerChatResponseProvider(id, provider, metadata) {
                checkProposedApiEnabled(extension, 'chatProvider');
                return extHostChatProvider.registerProvider(extension.identifier, id, provider, metadata);
            },
            registerSlashCommand(name, command, metadata) {
                checkProposedApiEnabled(extension, 'chatSlashCommands');
                return extHostChatSlashCommands.registerCommand(extension.identifier, name, command, metadata ?? { description: '' });
            },
            requestChatAccess(id) {
                checkProposedApiEnabled(extension, 'chatRequestAccess');
                return extHostChatProvider.requestChatResponseProvider(extension.identifier, id);
            },
            registerVariable(name, description, resolver) {
                checkProposedApiEnabled(extension, 'chatVariables');
                return extHostChatVariables.registerVariableResolver(extension, name, description, resolver);
            },
            registerMappedEditsProvider(selector, provider) {
                checkProposedApiEnabled(extension, 'mappedEditsProvider');
                return extHostLanguageFeatures.registerMappedEditsProvider(extension, selector, provider);
            },
            registerAgent(name, agent, metadata) {
                checkProposedApiEnabled(extension, 'chatAgents');
                return extHostChatAgents.registerAgent(extension.identifier, name, agent, metadata);
            }
        };
        return {
            version: initData.version,
            ai,
            authentication,
            commands,
            comments,
            chat,
            debug,
            env,
            extensions,
            interactive,
            l10n,
            languages,
            notebooks,
            scm,
            tasks,
            tests,
            window,
            workspace,
            Breakpoint: Breakpoint,
            TerminalOutputAnchor: TerminalOutputAnchor,
            ChatMessage: ChatMessage,
            ChatMessageRole: ChatMessageRole,
            ChatVariableLevel: ChatVariableLevel,
            CallHierarchyIncomingCall: CallHierarchyIncomingCall,
            CallHierarchyItem: CallHierarchyItem,
            CallHierarchyOutgoingCall: CallHierarchyOutgoingCall,
            CancellationError: errors.CancellationError,
            CancellationTokenSource: CancellationTokenSource,
            CandidatePortSource: CandidatePortSource,
            CodeAction: CodeAction,
            CodeActionKind: CodeActionKind,
            CodeActionTriggerKind: CodeActionTriggerKind,
            CodeLens: CodeLens,
            Color: Color,
            ColorInformation: ColorInformation,
            ColorPresentation: ColorPresentation,
            ColorThemeKind: ColorThemeKind,
            CommentMode: CommentMode,
            CommentState: CommentState,
            CommentThreadCollapsibleState: CommentThreadCollapsibleState,
            CommentThreadState: CommentThreadState,
            CompletionItem: CompletionItem,
            CompletionItemKind: CompletionItemKind,
            CompletionItemTag: CompletionItemTag,
            CompletionList: CompletionList,
            CompletionTriggerKind: CompletionTriggerKind,
            ConfigurationTarget: ConfigurationTarget,
            CustomExecution: CustomExecution,
            DebugAdapterExecutable: DebugAdapterExecutable,
            DebugAdapterInlineImplementation: DebugAdapterInlineImplementation,
            DebugAdapterNamedPipeServer: DebugAdapterNamedPipeServer,
            DebugAdapterServer: DebugAdapterServer,
            DebugConfigurationProviderTriggerKind: DebugConfigurationProviderTriggerKind,
            DebugConsoleMode: DebugConsoleMode,
            DecorationRangeBehavior: DecorationRangeBehavior,
            Diagnostic: Diagnostic,
            DiagnosticRelatedInformation: DiagnosticRelatedInformation,
            DiagnosticSeverity: DiagnosticSeverity,
            DiagnosticTag: DiagnosticTag,
            Disposable: Disposable,
            DocumentHighlight: DocumentHighlight,
            DocumentHighlightKind: DocumentHighlightKind,
            DocumentLink: DocumentLink,
            DocumentSymbol: DocumentSymbol,
            EndOfLine: EndOfLine,
            EnvironmentVariableMutatorType: EnvironmentVariableMutatorType,
            EvaluatableExpression: EvaluatableExpression,
            InlineValueText: InlineValueText,
            InlineValueVariableLookup: InlineValueVariableLookup,
            InlineValueEvaluatableExpression: InlineValueEvaluatableExpression,
            InlineCompletionTriggerKind: InlineCompletionTriggerKind,
            EventEmitter: Emitter,
            ExtensionKind: ExtensionKind,
            ExtensionMode: ExtensionMode,
            ExternalUriOpenerPriority: ExternalUriOpenerPriority,
            FileChangeType: FileChangeType,
            FileDecoration: FileDecoration,
            FileDecoration2: FileDecoration,
            FileSystemError: FileSystemError,
            FileType: files.FileType,
            FilePermission: files.FilePermission,
            FoldingRange: FoldingRange,
            FoldingRangeKind: FoldingRangeKind,
            FunctionBreakpoint: FunctionBreakpoint,
            InlineCompletionItem: InlineSuggestion,
            InlineCompletionList: InlineSuggestionList,
            Hover: Hover,
            IndentAction: languageConfiguration.IndentAction,
            Location: Location,
            MarkdownString: MarkdownString,
            OverviewRulerLane: OverviewRulerLane,
            ParameterInformation: ParameterInformation,
            PortAutoForwardAction: PortAutoForwardAction,
            Position: Position,
            ProcessExecution: ProcessExecution,
            ProgressLocation: ProgressLocation,
            QuickInputButtons: QuickInputButtons,
            Range: Range,
            RelativePattern: RelativePattern,
            Selection: Selection,
            SelectionRange: SelectionRange,
            SemanticTokens: SemanticTokens,
            SemanticTokensBuilder: SemanticTokensBuilder,
            SemanticTokensEdit: SemanticTokensEdit,
            SemanticTokensEdits: SemanticTokensEdits,
            SemanticTokensLegend: SemanticTokensLegend,
            ShellExecution: ShellExecution,
            ShellQuoting: ShellQuoting,
            SignatureHelp: SignatureHelp,
            SignatureHelpTriggerKind: SignatureHelpTriggerKind,
            SignatureInformation: SignatureInformation,
            SnippetString: SnippetString,
            SourceBreakpoint: SourceBreakpoint,
            StandardTokenType: StandardTokenType,
            StatusBarAlignment: StatusBarAlignment,
            SymbolInformation: SymbolInformation,
            SymbolKind: SymbolKind,
            SymbolTag: SymbolTag,
            Task: Task,
            TaskGroup: TaskGroup,
            TaskPanelKind: TaskPanelKind,
            TaskRevealKind: TaskRevealKind,
            TaskScope: TaskScope,
            TerminalLink: TerminalLink,
            TerminalQuickFixExecuteTerminalCommand: TerminalQuickFixCommand,
            TerminalQuickFixOpener: TerminalQuickFixOpener,
            TerminalLocation: TerminalLocation,
            TerminalProfile: TerminalProfile,
            TerminalExitReason: TerminalExitReason,
            TextDocumentSaveReason: TextDocumentSaveReason,
            TextEdit: TextEdit,
            SnippetTextEdit: SnippetTextEdit,
            TextEditorCursorStyle: TextEditorCursorStyle,
            TextEditorLineNumbersStyle: TextEditorLineNumbersStyle,
            TextEditorRevealType: TextEditorRevealType,
            TextEditorSelectionChangeKind: TextEditorSelectionChangeKind,
            SyntaxTokenType: SyntaxTokenType,
            TextDocumentChangeReason: TextDocumentChangeReason,
            ThemeColor: ThemeColor,
            ThemeIcon: ThemeIcon,
            TreeItem: TreeItem,
            TreeItemCheckboxState: TreeItemCheckboxState,
            TreeItemCollapsibleState: TreeItemCollapsibleState,
            TypeHierarchyItem: TypeHierarchyItem,
            UIKind: UIKind,
            Uri: URI,
            ViewColumn: ViewColumn,
            WorkspaceEdit: WorkspaceEdit,
            DocumentDropEdit: DocumentDropEdit,
            DocumentPasteEdit: DocumentPasteEdit,
            InlayHint: InlayHint,
            InlayHintLabelPart: InlayHintLabelPart,
            InlayHintKind: InlayHintKind,
            RemoteAuthorityResolverError: RemoteAuthorityResolverError,
            ResolvedAuthority: ResolvedAuthority,
            ManagedResolvedAuthority: ManagedResolvedAuthority,
            SourceControlInputBoxValidationType: SourceControlInputBoxValidationType,
            ExtensionRuntime: ExtensionRuntime,
            TimelineItem: TimelineItem,
            NotebookRange: NotebookRange,
            NotebookCellKind: NotebookCellKind,
            NotebookCellExecutionState: NotebookCellExecutionState,
            NotebookCellData: NotebookCellData,
            NotebookData: NotebookData,
            NotebookRendererScript: NotebookRendererScript,
            NotebookCellStatusBarAlignment: NotebookCellStatusBarAlignment,
            NotebookEditorRevealType: NotebookEditorRevealType,
            NotebookCellOutput: NotebookCellOutput,
            NotebookCellOutputItem: NotebookCellOutputItem,
            NotebookCellStatusBarItem: NotebookCellStatusBarItem,
            NotebookControllerAffinity: NotebookControllerAffinity,
            NotebookControllerAffinity2: NotebookControllerAffinity2,
            NotebookEdit: NotebookEdit,
            NotebookKernelSourceAction: NotebookKernelSourceAction,
            PortAttributes: PortAttributes,
            LinkedEditingRanges: LinkedEditingRanges,
            TestResultState: TestResultState,
            TestRunRequest: TestRunRequest,
            TestMessage: TestMessage,
            TestMessage2: TestMessage,
            TestTag: TestTag,
            TestRunProfileKind: TestRunProfileKind,
            TextSearchCompleteMessageType: TextSearchCompleteMessageType,
            DataTransfer: DataTransfer,
            DataTransferItem: DataTransferItem,
            CoveredCount: CoveredCount,
            FileCoverage: FileCoverage,
            StatementCoverage: StatementCoverage,
            BranchCoverage: BranchCoverage,
            FunctionCoverage: FunctionCoverage,
            WorkspaceTrustState: WorkspaceTrustState,
            LanguageStatusSeverity: LanguageStatusSeverity,
            QuickPickItemKind: QuickPickItemKind,
            InputBoxValidationSeverity: InputBoxValidationSeverity,
            TabInputText: TextTabInput,
            TabInputTextDiff: TextDiffTabInput,
            TabInputTextMerge: TextMergeTabInput,
            TabInputCustom: CustomEditorTabInput,
            TabInputNotebook: NotebookEditorTabInput,
            TabInputNotebookDiff: NotebookDiffEditorTabInput,
            TabInputWebview: WebviewEditorTabInput,
            TabInputTerminal: TerminalEditorTabInput,
            TabInputInteractiveWindow: InteractiveWindowInput,
            TelemetryTrustedValue: TelemetryTrustedValue,
            LogLevel: LogLevel,
            EditSessionIdentityMatch: EditSessionIdentityMatch,
            InteractiveSessionVoteDirection: InteractiveSessionVoteDirection,
            InteractiveSessionCopyKind: InteractiveSessionCopyKind,
            InteractiveEditorResponseFeedbackKind: InteractiveEditorResponseFeedbackKind,
            StackFrameFocus: StackFrameFocus,
            ThreadFocus: ThreadFocus,
            RelatedInformationType: RelatedInformationType
        };
    };
}
export { createApiFactoryAndRegisterActors };
