// vscode/vscode/src/vs/workbench/contrib/comments/browser/reactionsAction.js
import * as nls from 'monaco-editor/esm/vs/nls.js';
import * as dom from 'monaco-editor/esm/vs/base/browser/dom.js';
import { Action } from 'monaco-editor/esm/vs/base/common/actions.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import { ActionViewItem } from 'monaco-editor/esm/vs/base/browser/ui/actionbar/actionViewItems.js';
class ToggleReactionsAction extends Action {
    static { this.ID = 'toolbar.toggle.pickReactions'; }
    constructor(toggleDropdownMenu, title) {
        super(ToggleReactionsAction.ID, title || ( nls.localizeWithPath(
            'vs/workbench/contrib/comments/browser/reactionsAction',
            'pickReactions',
            "Pick Reactions..."
        )), 'toggle-reactions', true);
        this._menuActions = [];
        this.toggleDropdownMenu = toggleDropdownMenu;
    }
    run() {
        this.toggleDropdownMenu();
        return Promise.resolve(true);
    }
    get menuActions() {
        return this._menuActions;
    }
    set menuActions(actions) {
        this._menuActions = actions;
    }
}
class ReactionActionViewItem extends ActionViewItem {
    constructor(action) {
        super(null, action, {});
    }
    updateLabel() {
        if (!this.label) {
            return;
        }
        const action = this.action;
        if (action.class) {
            this.label.classList.add(action.class);
        }
        if (!action.icon) {
            const reactionLabel = dom.append(this.label, dom.$('span.reaction-label'));
            reactionLabel.innerText = action.label;
        }
        else {
            const reactionIcon = dom.append(this.label, dom.$('.reaction-icon'));
            const uri = URI.revive(action.icon);
            reactionIcon.style.backgroundImage = dom.asCSSUrl(uri);
            reactionIcon.title = action.label;
        }
        if (action.count) {
            const reactionCount = dom.append(this.label, dom.$('span.reaction-count'));
            reactionCount.innerText = `${action.count}`;
        }
    }
    getTooltip() {
        const action = this.action;
        const toggleMessage = action.enabled ? ( nls.localizeWithPath(
            'vs/workbench/contrib/comments/browser/reactionsAction',
            'comment.toggleableReaction',
            "Toggle reaction, "
        )) : '';
        if (action.count === undefined) {
            return ( nls.localizeWithPath('vs/workbench/contrib/comments/browser/reactionsAction', {
                key: 'comment.reactionLabelNone', comment: [
                    'This is a tooltip for an emoji button so that the current user can toggle their reaction to a comment.',
                    'The first arg is localized message "Toggle reaction" or empty if the user doesn\'t have permission to toggle the reaction, the second is the name of the reaction.'
                ]
            }, "{0}{1} reaction", toggleMessage, action.label));
        }
        else if (action.count === 1) {
            return ( nls.localizeWithPath('vs/workbench/contrib/comments/browser/reactionsAction', {
                key: 'comment.reactionLabelOne', comment: [
                    'This is a tooltip for an emoji that is a "reaction" to a comment where the count of the reactions is 1.',
                    'The emoji is also a button so that the current user can also toggle their own emoji reaction.',
                    'The first arg is localized message "Toggle reaction" or empty if the user doesn\'t have permission to toggle the reaction, the second is the name of the reaction.'
                ]
            }, "{0}1 reaction with {1}", toggleMessage, action.label));
        }
        else if (action.count > 1) {
            return ( nls.localizeWithPath('vs/workbench/contrib/comments/browser/reactionsAction', {
                key: 'comment.reactionLabelMany', comment: [
                    'This is a tooltip for an emoji that is a "reaction" to a comment where the count of the reactions is greater than 1.',
                    'The emoji is also a button so that the current user can also toggle their own emoji reaction.',
                    'The first arg is localized message "Toggle reaction" or empty if the user doesn\'t have permission to toggle the reaction, the second is number of users who have reacted with that reaction, and the third is the name of the reaction.'
                ]
            }, "{0}{1} reactions with {2}", toggleMessage, action.count, action.label));
        }
        return undefined;
    }
}
class ReactionAction extends Action {
    static { this.ID = 'toolbar.toggle.reaction'; }
    constructor(id, label = '', cssClass = '', enabled = true, actionCallback, icon, count) {
        super(ReactionAction.ID, label, cssClass, enabled, actionCallback);
        this.icon = icon;
        this.count = count;
    }
}
export { ReactionAction, ReactionActionViewItem, ToggleReactionsAction };
